{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6202ce00-100b-4ea0-8e0a-ef88bca29f6a",
   "metadata": {},
   "source": [
    "# GWAS in three cohorts\n",
    "Author: Jose Jaime Martinez-Magana\n",
    "Day: 04172025\n",
    "In this analysis an GWAS was performed using Regenie for three cohorts in the LAGC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c123618-2971-43f5-9a68-8678573d3bb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Perform a PCA\n",
    "# Request computational resources\n",
    "srun --pty --mem=128G --time=02:00:00 -p day bash\n",
    "# Upload PLINK for LD pruning\n",
    "module load PLINK\n",
    "\n",
    "## First, we will perform an LD pruning with plink\n",
    "in=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas\"\n",
    "# Perform first pass of LD pruning\n",
    "# MxGDAR-Fz1\n",
    "plink2 --pfile ${in}00mxgdarfz1/00databases/01genotypes/01forgwas/mxgdar1_merged\\\n",
    "    --indep-pairwise 50 5 0.2 --geno 0.01 --mind 0.01 --maf 0.05\\\n",
    "    --out ${in}00mxgdarfz1/00databases/01genotypes/00forpca/mxgdar1_merged_indep_snps\n",
    "plink2 --pfile ${in}00mxgdarfz1/00databases/01genotypes/01forgwas/mxgdar1_merged\\\n",
    "    --extract ${in}00mxgdarfz1/00databases/01genotypes/00forpca/mxgdar1_merged_indep_snps.prune.in --make-bed\\\n",
    "    --out ${in}00mxgdarfz1/00databases/01genotypes/00forpca/mxgdar1_merged_forpcair\n",
    "\n",
    "# MxGDAR-Fz2\n",
    "plink2 --pfile ${in}01mxgdarfz2/00databases/01genotypes/01forgwas/mxgdar2_merged\\\n",
    "    --indep-pairwise 50 5 0.2 --geno 0.01 --mind 0.01 --maf 0.05\\\n",
    "    --out ${in}01mxgdarfz2/00databases/01genotypes/00forpca/mxgdar2_merged_indep_snps\n",
    "plink2 --pfile ${in}01mxgdarfz2/00databases/01genotypes/01forgwas/mxgdar2_merged\\\n",
    "    --extract ${in}01mxgdarfz2/00databases/01genotypes/00forpca/mxgdar2_merged_indep_snps.prune.in --make-bed\\\n",
    "    --out ${in}01mxgdarfz2/00databases/01genotypes/00forpca/mxgdar2_merged_forpcair\n",
    "\n",
    "# BHRC\n",
    "plink2 --pfile ${in}02bhrc/00databases/01genotypes/01forgwas/bhrc_merged\\\n",
    "    --indep-pairwise 50 5 0.2 --geno 0.01 --mind 0.01 --maf 0.05\\\n",
    "    --out ${in}02bhrc/00databases/01genotypes/00forpca/bhrc_merged_indep_snps\n",
    "plink2 --pfile ${in}02bhrc/00databases/01genotypes/01forgwas/bhrc_merged\\\n",
    "    --extract ${in}02bhrc/00databases/01genotypes/00forpca/bhrc_merged_indep_snps.prune.in --make-bed\\\n",
    "    --out ${in}02bhrc/00databases/01genotypes/00forpca/bhrc_merged_forpcair"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efb5fc97-d874-4612-9470-d7e30ac7a088",
   "metadata": {},
   "source": [
    "### PCA with PCAiR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c643479c-1779-4be8-a966-25649b681d03",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH --job-name=pca_on_merged_data\n",
    "#SBATCH --out=\"slurm-%j.out\"\n",
    "#SBATCH --time=04:00:00\n",
    "#SBATCH --nodes=1 --ntasks=1\n",
    "#SBATCH --mem-per-cpu=128G\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --partition=day\n",
    "####################################################################################\n",
    "# Script for PCA without reference in several cohorts\n",
    "# day: 08 April 2025\n",
    "# analyzer: Jose Jaime Martinez-Magana\n",
    "# cluster: Grace - HPC Yale\n",
    "####################################################################################\n",
    "## Load modules to cluster\n",
    "# Load Genesis for PCA analysis\n",
    "module load miniconda;conda activate genesis_pca\n",
    "# Execute Rscript\n",
    "Rscript pca.R\n",
    "\n",
    "####################################################################################\n",
    "# Add this into a pca.R file\n",
    "####################################################################################\n",
    "# load libraries\n",
    "library(GENESIS)\n",
    "library(GWASTools)\n",
    "library(SNPRelate)\n",
    "library(SeqArray)\n",
    "library(parallel)\n",
    "library(BiocParallel);\n",
    "\n",
    "############################################################################################\n",
    "## !!! This is the only section that needs adjustment based on your data !!! ##\n",
    "## Set parameters\n",
    "# Change directory to the directory where the PLINK LD pruned files are stored\n",
    "wkdir=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/01genotypes/00forpca\"\n",
    "setwd(wkdir)\n",
    "# Create a list of files for input to PCAiR\n",
    "filename=\"mxgdar1_merged_forpcair\"\n",
    "bed_f=paste0(filename,\".bed\",sep=\"\")\n",
    "bim_f=paste0(filename,\".bim\",sep=\"\")\n",
    "fam_f=paste0(filename,\".fam\",sep=\"\")\n",
    "\n",
    "# Create output files for a GDS file, keep the GDS extension\n",
    "gds_output=\"mxgdar1_merged_forpcair.gds\"\n",
    "\n",
    "# Create a file name to store the PCA output, keep the CSV extension\n",
    "outfile_name=\"mxgdar1_pcs32_04122025.csv\"\n",
    "\n",
    "# create gds output files\n",
    "snpgdsBED2GDS(bed.fn=bed_f,\n",
    "              bim.fn=bim_f,\n",
    "              fam.fn=fam_f,\n",
    "              family=TRUE,\n",
    "              out.gdsfn=gds_output)\n",
    "####################################################################################\n",
    "# establishing a seed\n",
    "set.seed(1000)\n",
    "# establishing cores\n",
    "cores=detectCores()\n",
    "####################################################################################\n",
    "# Create functions\n",
    "# Create a function for LD pruning\n",
    "ld_prun = function(gds){\n",
    "snpset=snpgdsLDpruning(gds,\n",
    "                       method=\"corr\",\n",
    "                       slide.max.bp=10e6,\n",
    "                       ld.threshold=sqrt(0.1),\n",
    "                       maf=0.01,\n",
    "                       missing.rate=0.01,\n",
    "                       verbose=TRUE, num.thread=cores);\n",
    "    pruned=unlist(snpset,\n",
    "                  use.names=FALSE);\n",
    "return(pruned)\n",
    "}\n",
    "\n",
    "# Create a function for KING-robust analysis\n",
    "king_mat = function(gds){\n",
    "samp.id=read.gdsn(index.gdsn(gds, \"sample.id\"))\n",
    "ibd.robust=snpgdsIBDKING(gds, sample.id=samp.id, family.id=NULL, maf=0.01,missing.rate=0.01,num.thread=cores)\n",
    "return(ibd.robust)\n",
    "}\n",
    "\n",
    "# Create a function for PCAiR\n",
    "pcair_r = function(gds_geno, pruned, KINGmat){\n",
    "pcair=pcair(gds_geno, snp.include=pruned,\n",
    "            kinobj=KINGmat, divobj=KINGmat)\n",
    "return(pcair)\n",
    "}\n",
    "####################################################################################\n",
    "# set input data\n",
    "gds=gds_output\n",
    "\n",
    "# run analysis for PCs with relationships\n",
    "# open the GDS object\n",
    "gds=snpgdsOpen(gds);\n",
    "# LD pruning\n",
    "pruned=ld_prun(gds)\n",
    "# build KING matrix\n",
    "KINGmat=king_mat(gds)\n",
    "# adjust KING matrix\n",
    "KINGmat_m=KINGmat$kinship\n",
    "# add sampleIDs to colnames y row names\n",
    "colnames(KINGmat_m)=KINGmat$sample.id\n",
    "rownames(KINGmat_m)=KINGmat$sample.id\n",
    "# get samples in gds\n",
    "gds_samples=read.gdsn(index.gdsn(gds, \"sample.id\"))\n",
    "# close the gds object\n",
    "snpgdsClose(gds)\n",
    "####################################################################################\n",
    "# Warning: review this step!!!!!\n",
    "# read the GDS object\n",
    "gds=GdsGenotypeReader(filename=gds_output)\n",
    "\n",
    "# create a GenotypeData class object\n",
    "gds_geno=GenotypeData(gds)\n",
    "\n",
    "# run PCAiR\n",
    "PCair=pcair_r(gds_geno, pruned, KINGmat_m)\n",
    "\n",
    "# Create a file of PCs\n",
    "PCair_df=as.data.frame(PCair$vectors)\n",
    "# assign colnames\n",
    "colnames(PCair_df)=paste0(\"PC\",rep(1:32))\n",
    "# add IID\n",
    "PCair_df$IID=row.names(PCair_df)\n",
    "\n",
    "# save file\n",
    "write.csv(file=outfile_name,\n",
    "         PCair_df,\n",
    "         quote=FALSE,\n",
    "         row.names=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8072c429-ed11-4ec3-9045-04838832f6bc",
   "metadata": {},
   "source": [
    "### Running GWAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9370174f-dffb-4a93-bee7-e7b08f4d6066",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Recode pgen to bgen \n",
    "# The pgen file gave an error in the psam file\n",
    "module load PLINK\n",
    "plink2 --pfile bhrc_merged --export bgen-1.1 --snps-only 'just-acgt' --max-alleles 2 --out bhrc_merged --threads 30\n",
    "\n",
    "# Transform pgen to bfile for step1\n",
    "plink2\\\n",
    "    --pfile bhrc_merged\\\n",
    "    --maf 0.01 --mac 100 --geno 0.1 --hwe 1e-10 --mind 0.1\\\n",
    "    --write-snplist --write-samples --no-id-header --make-bed --snps-only 'just-acgt' --max-alleles 2\\\n",
    "    --out bhrc_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48869c60-92df-4817-9b19-29c073e40c85",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH --job-name=step1_bhrc\n",
    "#SBATCH --out=\"slurm-%j.out\"\n",
    "#SBATCH --time=6-23:00:00\n",
    "#SBATCH --nodes=1 --ntasks=1\n",
    "#SBATCH --mem-per-cpu=128G\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --partition=week\n",
    "####################################################################################\n",
    "# Script for GWAS of smoking traits in BHRC\n",
    "# day: 18 April 2025\n",
    "# analyzer: Jose Jaime Martinez-Magana\n",
    "# cluster: Grace - HPC Yale\n",
    "####################################################################################\n",
    "## Load modules to cluster\n",
    "# Load Regenie for GWAS analysis\n",
    "module load miniconda;conda activate regenie_env_v01202024\n",
    "\n",
    "## Set parameters\n",
    "# Set genotype data\n",
    "inge=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/01genotypes/01forgwas/bhrc_merged\"\n",
    "# Set phenotype and covariates files for females\n",
    "female_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_female_pheno_bt_forregenie_v04182025.txt\"\n",
    "female_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_female_pheno_qt_forregenie_v04182025.txt\"\n",
    "female_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_female_covar_forregenie_v04182025.txt\"\n",
    "# Set phenotype and covariates files for males\n",
    "male_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_male_pheno_bt_forregenie_v04182025.txt\"\n",
    "male_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_male_pheno_qt_forregenie_v04182025.txt\"\n",
    "male_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_male_covar_forregenie_v04182025.txt\"\n",
    "# Set output for step 1\n",
    "out_step1_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/00step1/out_step1_qt_female_04182025\"\n",
    "out_step1_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/00step1/out_step1_bt_female_04182025\"\n",
    "out_step1_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/00step1/out_step1_qt_male_04182025\"\n",
    "out_step1_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/00step1/out_step1_bt_female_04182025\"\n",
    "# Set output for step 2\n",
    "out_step2_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/01step2/out_step2_qt_female_04182025\"\n",
    "out_step2_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/01step2/out_step2_bt_female_04182025\"\n",
    "out_step2_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/01step2/out_step2_qt_female_04182025\"\n",
    "out_step2_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/01step2/out_step2_bt_female_04182025\"\n",
    "## Add zeros to the FID if you get errors. You could use the following code in bash\n",
    "# awk 'BEGIN {OFS=\"\\t\"} NR==1 {print \"FID\", $0; next} {print \"0\", $0}' \"${female_inph_qt}\" > tmpfile && mv tmpfile \"${female_inph_qt}\"\n",
    "\n",
    "## !!! Warning, we will use HapMap SNPs for Step 1!!!\n",
    "# Hapmap SNPs path\n",
    "hapmap=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/11references/w_hm3_hg38.snplist\"\n",
    "\n",
    "## Running Step1\n",
    "# Running for females for quantitative traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${female_inco}\\\n",
    "    --phenoFile ${female_inph_qt}\\\n",
    "    --bsize 10000\\\n",
    "    --qt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_qt_female}\n",
    "# Running for females for binary traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${female_inco}\\\n",
    "    --phenoFile ${female_inph_bt}\\\n",
    "    --bsize 10000\\\n",
    "    --iid-only\\\n",
    "    --bt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_bt_female}\n",
    "# Running for males for quantitative traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${male_inco}\\\n",
    "    --phenoFile ${male_inph_qt}\\\n",
    "    --bsize 10000\\\n",
    "    --iid-only\\\n",
    "    --qt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_qt_male}\n",
    "# Running for males for binary traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${male_inco}\\\n",
    "    --phenoFile ${male_inph_bt}\\\n",
    "    --bsize 10000\\\n",
    "    --iid-only\\\n",
    "    --bt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_bt_male}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e71bbb80-3135-4eed-b6a6-5fbf3c79e4af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### WARNING: using 1M SNPs takes a lot of time, reducing the number of SNPs to 250K\n",
    "# Hapmap SNPs path\n",
    "hapmap=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/11references/w_hm3_hg38.snplist\"\n",
    "shuf -n250000 ${hapmap} > w_hm3_hg38_random250K.snplist\n",
    "\n",
    "#### WARNING: running with 500K SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0a8f40f-b47d-4169-a777-e174be0488a5",
   "metadata": {},
   "source": [
    "### Step 1\n",
    "This was the script used for the step 1 of regenie"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc3932a0-7c3b-48d7-a412-ef6b43f96719",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH --job-name=step1_bhrc\n",
    "#SBATCH --out=\"slurm-%j.out\"\n",
    "#SBATCH --time=6-23:00:00\n",
    "#SBATCH --nodes=1 --ntasks=1\n",
    "#SBATCH --mem-per-cpu=128G\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --partition=week\n",
    "####################################################################################\n",
    "# Script for GWAS of smoking traits in BHRC\n",
    "# day: 18 April 2025\n",
    "# analyzer: Jose Jaime Martinez-Magana\n",
    "# cluster: Grace - HPC Yale\n",
    "####################################################################################\n",
    "## Load modules to cluster\n",
    "# Load Regenie for GWAS analysis\n",
    "module load miniconda;conda activate regenie_env_v01202024\n",
    "\n",
    "## Set parameters\n",
    "# Set genotype data\n",
    "inge=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/01genotypes/01forgwas/bhrc_merged\"\n",
    "# Set phenotype and covariates files for females\n",
    "female_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_female_pheno_bt_forregenie_v04182025.txt\"\n",
    "female_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_female_pheno_qt_forregenie_v04182025.txt\"\n",
    "female_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_female_covar_forregenie_v04182025.txt\"\n",
    "# Set phenotype and covariates files for males\n",
    "male_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_male_pheno_bt_forregenie_v04182025.txt\"\n",
    "male_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_male_pheno_qt_forregenie_v04182025.txt\"\n",
    "male_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_male_covar_forregenie_v04182025.txt\"\n",
    "# Set output for step1\n",
    "out_step1_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/00step1/out_step1_qt_female_04182025\"\n",
    "out_step1_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/00step1/out_step1_bt_female_04182025\"\n",
    "out_step1_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/00step1/out_step1_qt_male_04182025\"\n",
    "out_step1_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/00step1/out_step1_bt_male_04182025\"\n",
    "# Set output for step2\n",
    "out_step2_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/01step2/out_step2_qt_female_04182025\"\n",
    "out_step2_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/01step2/out_step2_bt_female_04182025\"\n",
    "out_step2_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/01step2/out_step2_qt_male_04182025\"\n",
    "out_step2_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/01step2/out_step2_bt_male_04182025\"\n",
    "## Add zeros to the FID if you get errors, you could use the following code in bash\n",
    "# awk 'BEGIN {OFS=\"\\t\"} NR==1 {print \"FID\", $0; next} {print \"0\", $0}' \"${female_inph_qt}\" > tmpfile && mv tmpfile \"${female_inph_qt}\"\n",
    "\n",
    "## !!! Warning, we will used HapMap SNPs for Step 1!!!\n",
    "# Hapmap SNPs path\n",
    "hapmap=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/11references/w_hm3_hg38_random250K.snplist\"\n",
    "\n",
    "## Running Step1\n",
    "# Running for females for quantitative traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${female_inco}\\\n",
    "    --phenoFile ${female_inph_qt}\\\n",
    "    --bsize 400\\\n",
    "    --qt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_qt_female}\n",
    "# Running for females for binary traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${female_inco}\\\n",
    "    --phenoFile ${female_inph_bt}\\\n",
    "    --bsize 400\\\n",
    "    --iid-only\\\n",
    "    --bt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_bt_female}\n",
    "# Running for males for quantitative traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${male_inco}\\\n",
    "    --phenoFile ${male_inph_qt}\\\n",
    "    --bsize 400\\\n",
    "    --iid-only\\\n",
    "    --qt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_qt_male}\n",
    "# Running for males for binary traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${male_inco}\\\n",
    "    --phenoFile ${male_inph_bt}\\\n",
    "    --bsize 400\\\n",
    "    --iid-only\\\n",
    "    --bt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_bt_male}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ed2da9d-26d5-458a-8d28-e3951cdd5fa1",
   "metadata": {},
   "source": [
    "## Step 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a77eb9e-5765-4750-8708-64863aca6b1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH --job-name=step2_bhrc\n",
    "#SBATCH --out=\"slurm-%j.out\"\n",
    "#SBATCH --time=6-23:00:00\n",
    "#SBATCH --nodes=1 --ntasks=1\n",
    "#SBATCH --mem-per-cpu=128G\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --partition=week\n",
    "####################################################################################\n",
    "# Script for GWAS of smoking traits in BHRC\n",
    "# day: 18 April 2025\n",
    "# analyzer: Jose Jaime Martinez-Magana\n",
    "# cluster: Grace - HPC Yale\n",
    "####################################################################################\n",
    "## Load modules to cluster\n",
    "# Load Regenie for GWAS analysis\n",
    "module load miniconda;conda activate regenie_env_v01202024\n",
    "\n",
    "## Set parameters\n",
    "# Set genotype data\n",
    "inge=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/01genotypes/01forgwas/bhrc_merged\"\n",
    "# Set phenotype and covariates files for females\n",
    "female_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_female_pheno_bt_forregenie_v04182025.txt\"\n",
    "female_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_female_pheno_qt_forregenie_v04182025.txt\"\n",
    "female_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_female_covar_forregenie_v04182025.txt\"\n",
    "# Set phenotype and covariates files for males\n",
    "male_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_male_pheno_bt_forregenie_v04182025.txt\"\n",
    "male_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_male_pheno_qt_forregenie_v04182025.txt\"\n",
    "male_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/00databases/00phenotype/bhrc_pheno_male_covar_forregenie_v04182025.txt\"\n",
    "# Set output for step1\n",
    "out_step1_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/00step1/out_step1_qt_female_04182025\"\n",
    "out_step1_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/00step1/out_step1_bt_female_04182025\"\n",
    "out_step1_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/00step1/out_step1_qt_male_04182025\"\n",
    "out_step1_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/00step1/out_step1_bt_male_04182025\"\n",
    "# Set output for step2\n",
    "out_step2_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/01step2/out_step2_qt_female_04182025\"\n",
    "out_step2_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/01step2/out_step2_bt_female_04182025\"\n",
    "out_step2_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/01step2/out_step2_qt_male_04182025\"\n",
    "out_step2_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/02bhrc/01results/01step2/out_step2_bt_male_04182025\"\n",
    "\n",
    "# Running GWAS for females \n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${female_inph_qt} \\\n",
    "  --covarFile ${female_inco} \\\n",
    "  --qt \\\n",
    "  --pred ${out_step1_qt_female}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_qt_female}\n",
    "\n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${female_inph_bt} \\\n",
    "  --covarFile ${female_inco} \\\n",
    "  --bt \\\n",
    "  --firth --approx --pThresh 0.01 \\\n",
    "  --pred ${out_step1_qt_female}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_qt_female}\n",
    "\n",
    "# Running GWAS for males \n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${male_inph_qt} \\\n",
    "  --covarFile ${male_inco} \\\n",
    "  --qt \\\n",
    "  --pred ${out_step1_qt_male}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_qt_male}\n",
    "\n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${male_inph_bt} \\\n",
    "  --covarFile ${male_inco} \\\n",
    "  --bt \\\n",
    "  --firth --approx --pThresh 0.01 \\\n",
    "  --pred ${out_step1_qt_male}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_qt_male}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78fe1f77-c761-45fe-b7a8-af57bf23b87a",
   "metadata": {},
   "source": [
    "# Code for MXGDARfz1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5663d34-ce2a-434e-8cb7-f53b1a1136c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH --job-name=step1_mxgdar1\n",
    "#SBATCH --out=\"slurm-%j.out\"\n",
    "#SBATCH --time=6-23:00:00\n",
    "#SBATCH --nodes=1 --ntasks=1\n",
    "#SBATCH --mem-per-cpu=128G\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --partition=week\n",
    "####################################################################################\n",
    "# Script for GWAS of smoking traits in MxGDAR-fz1\n",
    "# day: 18 April 2025\n",
    "# analyzer: Jose Jaime Martinez-Magana\n",
    "# cluster: Grace - HPC Yale\n",
    "####################################################################################\n",
    "## Load modules to cluster\n",
    "# Load Regenie for GWAS analysis\n",
    "module load miniconda;conda activate regenie_env_v01202024\n",
    "\n",
    "## Set parameters\n",
    "# Set genotype data\n",
    "inge=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/01genotypes/01forgwas/mxgdar1_merged\"\n",
    "# Set phenotype and covariates files for females\n",
    "female_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/00phenotype/mxgdarfz1_pheno_female_pheno_bt_forregenie_v04202025.txt\"\n",
    "female_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/00phenotype/mxgdarfz1_pheno_female_pheno_qt_forregenie_v04202025.txt\"\n",
    "female_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/00phenotype/mxgdarfz1_pheno_female_covar_forregenie_v04202025.txt\"\n",
    "# Set phenotype and covariates files for males\n",
    "male_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/00phenotype/mxgdarfz1_pheno_male_pheno_bt_forregenie_v04202025.txt\"\n",
    "male_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/00phenotype/mxgdarfz1_pheno_male_pheno_qt_forregenie_v04202025.txt\"\n",
    "male_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/00phenotype/mxgdarfz1_pheno_male_covar_forregenie_v04202025.txt\"\n",
    "# Set output for step1\n",
    "out_step1_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/00step1/out_step1_qt_female_04202025\"\n",
    "out_step1_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/00step1/out_step1_bt_female_04202025\"\n",
    "out_step1_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/00step1/out_step1_qt_male_04202025\"\n",
    "out_step1_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/00step1/out_step1_bt_male_04202025\"\n",
    "# Set output for step2\n",
    "out_step2_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/01step2/out_step2_qt_female_04202025\"\n",
    "out_step2_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/01step2/out_step2_bt_female_04202025\"\n",
    "out_step2_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/01step2/out_step2_qt_male_04202025\"\n",
    "out_step2_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/01step2/out_step2_bt_male_04202025\"\n",
    "## Add zeros to the FID if you get errors, you could use the following code in bash\n",
    "# awk 'BEGIN {OFS=\"\\t\"} NR==1 {print \"FID\", $0; next} {print \"0\", $0}' \"${female_inph_qt}\" > tmpfile && mv tmpfile \"${female_inph_qt}\"\n",
    "\n",
    "## !!! Warning, we will used HapMap SNPs for Step 1!!!\n",
    "# Hapmap SNPs path\n",
    "hapmap=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/11references/w_hm3_hg38_random250K.snplist\"\n",
    "\n",
    "## Running Step1\n",
    "# Running for females for quantitative traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${female_inco}\\\n",
    "    --phenoFile ${female_inph_qt}\\\n",
    "    --bsize 400\\\n",
    "    --qt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_qt_female}\n",
    "# Running for females for binary traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${female_inco}\\\n",
    "    --phenoFile ${female_inph_bt}\\\n",
    "    --bsize 400\\\n",
    "    --iid-only\\\n",
    "    --bt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_bt_female}\n",
    "# Running for males for quantitative traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${male_inco}\\\n",
    "    --phenoFile ${male_inph_qt}\\\n",
    "    --bsize 400\\\n",
    "    --iid-only\\\n",
    "    --qt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_qt_male}\n",
    "# Running for males for binary traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${male_inco}\\\n",
    "    --phenoFile ${male_inph_bt}\\\n",
    "    --bsize 400\\\n",
    "    --iid-only\\\n",
    "    --bt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_bt_male}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d89faf4-b56f-42d4-a037-98b21fe7affc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH --job-name=step2_mxgdar1\n",
    "#SBATCH --out=\"slurm-%j.out\"\n",
    "#SBATCH --time=6-23:00:00\n",
    "#SBATCH --nodes=1 --ntasks=1\n",
    "#SBATCH --mem-per-cpu=128G\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --partition=week\n",
    "####################################################################################\n",
    "# Script for GWAS of smoking traits in MxGDAR-fz1\n",
    "# day: 18 April 2025\n",
    "# analyzer: Jose Jaime Martinez-Magana\n",
    "# cluster: Grace - HPC Yale\n",
    "####################################################################################\n",
    "## Load modules to cluster\n",
    "# Load Regenie for GWAS analysis\n",
    "module load miniconda;conda activate regenie_env_v01202024\n",
    "\n",
    "## Set parameters\n",
    "# Set genotype data\n",
    "inge=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/01genotypes/01forgwas/mxgdar1_merged\"\n",
    "# Set phenotype and covariates files for females\n",
    "female_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/00phenotype/mxgdarfz1_pheno_female_pheno_bt_forregenie_v04202025.txt\"\n",
    "female_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/00phenotype/mxgdarfz1_pheno_female_pheno_qt_forregenie_v04202025.txt\"\n",
    "female_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/00phenotype/mxgdarfz1_pheno_female_covar_forregenie_v04202025.txt\"\n",
    "# Set phenotype and covariates files for males\n",
    "male_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/00phenotype/mxgdarfz1_pheno_male_pheno_bt_forregenie_v04202025.txt\"\n",
    "male_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/00phenotype/mxgdarfz1_pheno_male_pheno_qt_forregenie_v04202025.txt\"\n",
    "male_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/00databases/00phenotype/mxgdarfz1_pheno_male_covar_forregenie_v04202025.txt\"\n",
    "# Set output for step1\n",
    "out_step1_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/00step1/out_step1_qt_female_04202025\"\n",
    "out_step1_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/00step1/out_step1_bt_female_04202025\"\n",
    "out_step1_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/00step1/out_step1_qt_male_04202025\"\n",
    "out_step1_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/00step1/out_step1_bt_male_04202025\"\n",
    "# Set output for step2\n",
    "out_step2_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/01step2/out_step2_qt_female_04202025\"\n",
    "out_step2_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/01step2/out_step2_bt_female_04202025\"\n",
    "out_step2_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/01step2/out_step2_qt_male_04202025\"\n",
    "out_step2_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/00mxgdarfz1/01results/01step2/out_step2_bt_male_04202025\"\n",
    "\n",
    "# Running GWAS for females \n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${female_inph_qt} \\\n",
    "  --covarFile ${female_inco} \\\n",
    "  --qt \\\n",
    "  --pred ${out_step1_qt_female}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_qt_female}\n",
    "\n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${female_inph_bt} \\\n",
    "  --covarFile ${female_inco} \\\n",
    "  --bt \\\n",
    "  --firth --approx --pThresh 0.01 \\\n",
    "  --pred ${out_step1_bt_female}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_bt_female}\n",
    "\n",
    "# Running GWAS for males \n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${male_inph_qt} \\\n",
    "  --covarFile ${male_inco} \\\n",
    "  --qt \\\n",
    "  --pred ${out_step1_qt_male}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_qt_male}\n",
    "\n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${male_inph_bt} \\\n",
    "  --covarFile ${male_inco} \\\n",
    "  --bt \\\n",
    "  --firth --approx --pThresh 0.01 \\\n",
    "  --pred ${out_step1_bt_male}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_bt_male}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b1495f7-0f89-456b-b6d1-716ed64fdf1b",
   "metadata": {},
   "source": [
    "# MxGDARfz2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a3b1116-0a90-4793-a747-04dc352b463b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH --job-name=step1_mxgdar2\n",
    "#SBATCH --out=\"slurm-%j.out\"\n",
    "#SBATCH --time=6-23:00:00\n",
    "#SBATCH --nodes=1 --ntasks=1\n",
    "#SBATCH --mem-per-cpu=128G\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --partition=week\n",
    "####################################################################################\n",
    "# Script for GWAS of smoking traits in MxGDAR-fz2\n",
    "# day: 18 April 2025\n",
    "# analyzer: Jose Jaime Martinez-Magana\n",
    "# cluster: Grace - HPC Yale\n",
    "####################################################################################\n",
    "## Load modules to cluster\n",
    "# Load Regenie for GWAS analysis\n",
    "module load miniconda;conda activate regenie_env_v01202024\n",
    "\n",
    "## Set parameters\n",
    "# Set genotype data\n",
    "inge=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/01genotypes/01forgwas/mxgdar2_merged\"\n",
    "# Set phenotype and covariates files for females\n",
    "female_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/00phenotype/mxgdarfz2_pheno_female_pheno_bt_forregenie_v04202025.txt\"\n",
    "female_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/00phenotype/mxgdarfz2_pheno_female_pheno_qt_forregenie_v04202025.txt\"\n",
    "female_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/00phenotype/mxgdarfz2_pheno_female_covar_forregenie_v04202025.txt\"\n",
    "# Set phenotype and covariates files for males\n",
    "male_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/00phenotype/mxgdarfz2_pheno_male_pheno_bt_forregenie_v04202025.txt\"\n",
    "male_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/00phenotype/mxgdarfz2_pheno_male_pheno_qt_forregenie_v04202025.txt\"\n",
    "male_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/00phenotype/mxgdarfz2_pheno_male_covar_forregenie_v04202025.txt\"\n",
    "# Set output for step1\n",
    "out_step1_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/00step1/out_step1_qt_female_04202025\"\n",
    "out_step1_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/00step1/out_step1_bt_female_04202025\"\n",
    "out_step1_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/00step1/out_step1_qt_male_04202025\"\n",
    "out_step1_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/00step1/out_step1_bt_male_04202025\"\n",
    "# Set output for step2\n",
    "out_step2_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/01step2/out_step2_qt_female_04202025\"\n",
    "out_step2_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/01step2/out_step2_bt_female_04202025\"\n",
    "out_step2_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/01step2/out_step2_qt_male_04202025\"\n",
    "out_step2_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/01step2/out_step2_bt_male_04202025\"\n",
    "## Add zeros to the FID if you get errors. You could use the following code in bash\n",
    "# awk 'BEGIN {OFS=\"\\t\"} NR==1 {print \"FID\", $0; next} {print \"0\", $0}' \"${female_inph_qt}\" > tmpfile && mv tmpfile \"${female_inph_qt}\"\n",
    "\n",
    "## !!! Warning, we will used HapMap SNPs for Step 1!!!\n",
    "# Hapmap SNPs path\n",
    "hapmap=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/11references/w_hm3_hg38_random250K.snplist\"\n",
    "\n",
    "## Running Step1\n",
    "# Running for females for quantitative traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${female_inco}\\\n",
    "    --phenoFile ${female_inph_qt}\\\n",
    "    --bsize 400\\\n",
    "    --qt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_qt_female}\n",
    "# Running for females for binary traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${female_inco}\\\n",
    "    --phenoFile ${female_inph_bt}\\\n",
    "    --bsize 400\\\n",
    "    --iid-only\\\n",
    "    --bt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_bt_female}\n",
    "# Running for males for quantitative traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${male_inco}\\\n",
    "    --phenoFile ${male_inph_qt}\\\n",
    "    --bsize 400\\\n",
    "    --iid-only\\\n",
    "    --qt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_qt_male}\n",
    "# Running for males for binary traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${male_inco}\\\n",
    "    --phenoFile ${male_inph_bt}\\\n",
    "    --bsize 400\\\n",
    "    --iid-only\\\n",
    "    --bt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_bt_male}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88b2d18a-2af2-4030-a026-b6b628b3b74c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH --job-name=step2_mxgdar2\n",
    "#SBATCH --out=\"slurm-%j.out\"\n",
    "#SBATCH --time=6-23:00:00\n",
    "#SBATCH --nodes=1 --ntasks=1\n",
    "#SBATCH --mem-per-cpu=32G\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --partition=week\n",
    "####################################################################################\n",
    "# Script for GWAS of smoking traits in MxGDAR-fz2\n",
    "# day: 18 April 2025\n",
    "# analyzer: Jose Jaime Martinez-Magana\n",
    "# cluster: Grace - HPC Yale\n",
    "####################################################################################\n",
    "## Load modules to cluster\n",
    "# Load Regenie for GWAS analysis\n",
    "module load miniconda;conda activate regenie_env_v01202024\n",
    "\n",
    "## Set parameters\n",
    "# Set genotype data\n",
    "inge=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/01genotypes/01forgwas/mxgdar2_merged\"\n",
    "# Set phenotype and covariates files for females\n",
    "female_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/00phenotype/mxgdarfz2_pheno_female_pheno_bt_forregenie_v04202025.txt\"\n",
    "female_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/00phenotype/mxgdarfz2_pheno_female_pheno_qt_forregenie_v04202025.txt\"\n",
    "female_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/00phenotype/mxgdarfz2_pheno_female_covar_forregenie_v04202025.txt\"\n",
    "# Set phenotype and covariates files for males\n",
    "male_inph_bt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/00phenotype/mxgdarfz2_pheno_male_pheno_bt_forregenie_v04202025.txt\"\n",
    "male_inph_qt=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/00phenotype/mxgdarfz2_pheno_male_pheno_qt_forregenie_v04202025.txt\"\n",
    "male_inco=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/00databases/00phenotype/mxgdarfz2_pheno_male_covar_forregenie_v04202025.txt\"\n",
    "# Set output for step1\n",
    "out_step1_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/00step1/out_step1_qt_female_04202025\"\n",
    "out_step1_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/00step1/out_step1_bt_female_04202025\"\n",
    "out_step1_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/00step1/out_step1_qt_male_04202025\"\n",
    "out_step1_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/00step1/out_step1_bt_male_04202025\"\n",
    "# Set output for step2\n",
    "out_step2_qt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/01step2/out_step2_qt_female_04202025\"\n",
    "out_step2_bt_female=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/01step2/out_step2_bt_female_04202025\"\n",
    "out_step2_qt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/01step2/out_step2_qt_male_04202025\"\n",
    "out_step2_bt_male=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/01mxgdarfz2/01results/01step2/out_step2_bt_male_04202025\"\n",
    "\n",
    "# Running GWAS for females \n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${female_inph_qt} \\\n",
    "  --covarFile ${female_inco} \\\n",
    "  --qt \\\n",
    "  --pred ${out_step1_qt_female}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_qt_female}\n",
    "\n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${female_inph_bt} \\\n",
    "  --covarFile ${female_inco} \\\n",
    "  --bt \\\n",
    "  --firth --approx --pThresh 0.01 \\\n",
    "  --pred ${out_step1_bt_female}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_bt_female}\n",
    "\n",
    "# Running GWAS for males \n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${male_inph_qt} \\\n",
    "  --covarFile ${male_inco} \\\n",
    "  --qt \\\n",
    "  --pred ${out_step1_qt_male}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_qt_male}\n",
    "\n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${male_inph_bt} \\\n",
    "  --covarFile ${male_inco} \\\n",
    "  --bt \\\n",
    "  --firth --approx --pThresh 0.01 \\\n",
    "  --pred ${out_step1_bt_male}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_bt_male}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
