{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0f9dcb30-1a43-456d-a6aa-07f1511fdd38",
   "metadata": {},
   "source": [
    "# Inferencia de Componentes Principales\n",
    "\n",
    "Autor: Rafaella Ormond y Jose Jaime Martinez-Magana\n",
    "\n",
    "Este script realiza un Análisis de Componentes Principales (PCA) para generar covariables para el GWAS de rasgos de tabaquismo en las cohortes LAGC.\n",
    "\n",
    "**ENTRADA:** Archivos binarios de PLINK2 `(.pgen/.psan/.pvar)` + genoma de referencia `(.fa)` → control de calidad y conversión de formato\n",
    "**SALIDA:** Archivos VCF controlados por calidad, divididos por cromosoma (chr1-22), listos para imputación en los servidores TOPMed/Michigan\n",
    "\n",
    "### ***Requisitos:***\n",
    "### Descargar Plink\n",
    "Podemos descargar Plink versión 1.9 y versión 2.0 siguiendo los pasos de su sitio web.<br>\n",
    "Para instalar plink2 [acceda aquí](https://www.cog-genomics.org/plink/2.0/)<br>\n",
    "Para instalar plink1.9 [acceda aquí](https://www.cog-genomics.org/plink/1.9/) <br>\n",
    "\n",
    "### Descargar e instalar R\n",
    "\n",
    "Puede descargar e instalar R desde el Comprehensive R Archive Network (CRAN) para su sistema operativo:\n",
    "\n",
    "- [R para Windows](https://cran.r-project.org/bin/windows/base/)\n",
    "- [R para macOS](https://cran.r-project.org/bin/macosx/)\n",
    "- [R para Linux](https://cran.r-project.org/bin/linux/)\n",
    "\n",
    "Después de instalar R, puede instalar los paquetes requeridos como se muestra a continuación.\n",
    "\n",
    "### Descargar paquetes de R\n",
    "Instale los siguientes paquetes de R (haga clic en el nombre para visitar la página web de cada paquete):\n",
    "- [GENESIS](https://bioconductor.org/packages/release/bioc/html/GENESIS.html)\n",
    "- [GWASTools](https://bioconductor.org/packages/release/bioc/html/GWASTools.html)\n",
    "- [SNPRelate](https://bioconductor.org/packages/release/bioc/html/SNPRelate.html)\n",
    "- [SeqArray](https://bioconductor.org/packages/release/bioc/html/SeqArray.html)\n",
    "- [parallel](https://stat.ethz.ch/R-manual/R-devel/library/parallel/doc/parallel.pdf) (paquete incorporado en R)\n",
    "- [BiocParallel](https://bioconductor.org/packages/release/bioc/html/BiocParallel.html)\n",
    "- [stringi](https://cran.r-project.org/package=stringi)\n",
    "\n",
    "**Nota:** La mayoría de los paquetes de Bioconductor se pueden instalar usando los siguientes comandos en R:\n",
    "\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "\n",
    "BiocManager::install(c(\"GENESIS\", \"GWASTools\", \"SNPRelate\", \"SeqArray\", \"BiocParallel\"))\n",
    "install.packages(\"stringi\")\n",
    "\n",
    "### Pasos del análisis:\n",
    "1) Poda de LD usando PLINK\n",
    "2) Creación de archivos GDS<br>\n",
    "    2.1. Creación del script para crear los archivos gds<br>\n",
    "    2.2 Crear los archivos gds\n",
    "3) Cálculo de Componentes Principales usando PC-AiR\n",
    "   3.1. Desarrollo del script para PCA y GRM con PCAiR y PCRelate<br>\n",
    "   3.2. Generar los archivos de PCA y GRM<br>\n",
    "   3.3. Guardar el archivo de PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "902f7712-92fd-4c51-9aee-a65630999760",
   "metadata": {},
   "source": [
    "### 1. Pruning de LD usando PLINK\n",
    "\n",
    "***Descripción:***\n",
    "Este código es para realizar pruning de LD usando PLINK2 para el análisis de componentes principales.\n",
    "\n",
    "El formato plink2 utiliza pfiles, también se puede hacer la transformación para bfiles, pero si los archivos están divididos por cromosomas, con pfiles es más rápido unirlos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bff79d5a-5f88-433b-ad99-9108cde47cd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## First, we will perform an LD pruning with plink2\n",
    "# Substitute for the path and file name prefix of plink2 files\n",
    "in=\"/path_to_your_data/plink_prefix\"\n",
    "# Perform first pass of LD pruning\n",
    "plink2 --pfile ${in}\\\n",
    "    --indep-pairwise 50 5 0.2 --geno 0.01 --mind 0.01 --maf 0.05\\\n",
    "    --out ${in}_indep_snps\n",
    "plink2 --pfile ${in}\\\n",
    "    --extract ${in}_indep_snps.prune.in --make-bed\\\n",
    "    --out ${in}_forpcair"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55c51878-2c76-49b1-99a9-35edffb41270",
   "metadata": {},
   "source": [
    "### 2. Creación de archivos GDS\n",
    "\n",
    "***Descripción:***\n",
    "\n",
    "Este script creará los archivos gds que se usarán para realizar el cálculo de Componentes Principales en el siguiente paso."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47db38e0-d34e-4a78-8888-4e4fc38de01b",
   "metadata": {},
   "source": [
    "### 2.1 Creación del script para crear los archivos gds\n",
    "Nombre del script: **create_gds.Rscript**\n",
    "\n",
    "Puede descargar este script en github [ENLACE AQUÍ](https://github.com/ormondr/Smoking_GWAS_LAGC/blob/main/English/01PC/create_gds.Rscript)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd9f3135-8e24-46d5-9273-37ea7ad5d9a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################################################\n",
    "# This script uses a plink file to create a GDS file for input to PCAiR\n",
    "####################################################################################\n",
    "# set parameters\n",
    "# this function uses the library optparse to add arguments to the script\n",
    "# adding arguments to the script\n",
    "library(optparse) \n",
    "option_list = list(\n",
    "    make_option(c(\"--plinkfile\"), type=\"character\", default=NULL,\n",
    "                help=\"path to the directory storing the plink files, without plink extensions *bed/*bim/*fam\", metavar=\"character\"),\n",
    "    make_option(c(\"--out\"), type=\"character\", default=NULL,\n",
    "                help=\"output file name for GDS file, use complete path, example /data/analsyis/analysis.gds\", metavar=\"character\")\n",
    ");\n",
    "\n",
    "opt_parser = OptionParser(option_list=option_list);\n",
    "opt = parse_args(opt_parser);\n",
    "\n",
    "if (is.null(opt$plinkfile)){\n",
    "    print_help(opt_parser)\n",
    "    stop(\"At least one argument must be supplied (input file)\", call.=FALSE)\n",
    "}\n",
    "\n",
    "###################################################################################\n",
    "# load libraries\n",
    "library(GENESIS)\n",
    "library(GWASTools)\n",
    "library(SNPRelate)\n",
    "\n",
    "# create list of files\n",
    "bed_f=paste0(opt$plinkfile,\".bed\",sep=\"\")\n",
    "bim_f=paste0(opt$plinkfile,\".bim\",sep=\"\")\n",
    "fam_f=paste0(opt$plinkfile,\".fam\",sep=\"\")\n",
    "\n",
    "# create gds output files\n",
    "snpgdsBED2GDS(bed.fn=bed_f,\n",
    "              bim.fn=bim_f,\n",
    "              fam.fn=fam_f,\n",
    "              family=TRUE,\n",
    "              out.gdsfn=opt$out)\n",
    "# end of script"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d28a16f7-0640-4db5-8b8a-e8c3a79bf6f0",
   "metadata": {},
   "source": [
    "### 2.2 Crear los archivos gds\n",
    "***Descripción:***\n",
    "Este script ejecutará el script \"create_gds.Rscript\" creado en el paso anterior.<br>\n",
    "Obs: necesita modificar los parámetros para sus archivos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "641df291-dcad-4c15-bfda-2e6fc0a2d87f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating GDS files\n",
    "# set parameters\n",
    "Rscript='/path_to_your_data/create_gds.Rscript'\n",
    "inpath='/path_to_your_data/01plinkldpruned'\n",
    "oupath='/path_to_your_data/02gdsprunned'\n",
    "\n",
    "# create job list file\n",
    "do Rscript ${Rscript} --plinkfile=${inpath}/$cohort_name_hapmap_metalprs_allchrs --out=${oupath}/cohort_name_hapmap_metalprs_allchrs.gds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efb5fc97-d874-4612-9470-d7e30ac7a088",
   "metadata": {},
   "source": [
    "### 3. Cálculo de Componentes Principales usando PC-AiR\n",
    "\n",
    "**Descripción:**\n",
    "\n",
    "Este script realiza el Análisis de Componentes Principales (PCA), que se utilizará como covariables en el GWAS.\n",
    "El PCA se realiza usando la función `PC-AiR` del paquete **GENESIS**.\n",
    "\n",
    "- Para aprender más sobre PC-AiR y PC-Relate, consulte esta viñeta: [Guía PCA GENESIS](https://bioconductor.org/packages/devel/bioc/vignettes/GENESIS/inst/doc/pcair.html)\n",
    "  - Ver **Sección 3** para PC-AiR y **Sección 4** para PC-Relate.\n",
    "\n",
    "> **Nota:**\n",
    "> El PCA requiere poda de LD previa, que se realiza en el primer paso de este flujo de trabajo.\n",
    "> El archivo que contiene los 10 principales componentes se usará para todos los métodos GWAS (por ejemplo, Regenie, GMMAT, Saige).\n",
    "> El archivo GRM solo es necesario para GMMAT, pero recomendamos generarlo de todos modos — así no tendrá que repetir este paso si decide usar GMMAT más adelante."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53330f39-6a90-46b6-a60e-f9c0a1f7e1ab",
   "metadata": {},
   "source": [
    "### 3.1. Desarrollo del script para PCA y GRM con PCAiR y PCRelate\n",
    "*Script escrito en R.*\n",
    "Desarrollo del script para PCA y GRM con PCAiR y PCRelate <br>\n",
    "Este script no necesita ser modificado, las rutas y nombres de archivos se agregarán en el siguiente paso<br>\n",
    "Nombre del script: **create_grm.Rscript**\n",
    "\n",
    "Puede descargar este script en github [ENLACE AQUÍ](https://github.com/ormondr/Smoking_GWAS_LAGC/blob/main/English/01PC/create_grm.Rscript)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e61253a-6303-460c-8bbc-0026c82498ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################################################\n",
    "# R script for creating GRM and PCs\n",
    "# day: 25 July 2025\n",
    "# author: Rafaella Ormond and Jose Jaime Martinez-Magana\n",
    "####################################################################################\n",
    "# This script uses a gds file to estimate PCs and GRM using GENESIS\n",
    "####################################################################################\n",
    "# set parameters\n",
    "# this function uses the library optparse to add arguments to the script\n",
    "# adding arguments to the script\n",
    "library(optparse) \n",
    "option_list = list(\n",
    "    make_option(c(\"--gdsfile\"), type=\"character\", default=NULL,\n",
    "                help=\"path to gds file\", metavar=\"character\"),\n",
    "    make_option(c(\"--out\"), type=\"character\", default=NULL,\n",
    "                help=\"output path for PCAiR objects, use a complete directory path and a potential name of files. Example: /home/user/out_pca\", metavar=\"character\")\n",
    ");\n",
    "\n",
    "opt_parser = OptionParser(option_list=option_list);\n",
    "opt = parse_args(opt_parser);\n",
    "\n",
    "if (is.null(opt$gdsfile)){\n",
    "  print_help(opt_parser)\n",
    "  stop(\"At least one argument must be supplied (input file)\", call.=FALSE)\n",
    "}\n",
    "\n",
    "###################################################################################\n",
    "# load libraries\n",
    "library(GENESIS)\n",
    "library(GWASTools)\n",
    "library(SNPRelate)\n",
    "library(SeqArray)\n",
    "library(parallel)\n",
    "library(BiocParallel)\n",
    "library(stringi);\n",
    "\n",
    "# stablishing a seed\n",
    "set.seed(1000)\n",
    "# stablishing cores\n",
    "cores=detectCores()\n",
    "\n",
    "# create function for LD prunning\n",
    "ld_prun=function(gds){\n",
    "snpset=snpgdsLDpruning(gds,\n",
    "                       method=\"corr\",\n",
    "                       slide.max.bp=10e6,\n",
    "                       ld.threshold=sqrt(0.1),\n",
    "                       maf=0.01,\n",
    "                       missing.rate=0.01,\n",
    "                       verbose=TRUE, num.thread=cores);\n",
    "    pruned=unlist(snpset,\n",
    "                  use.names=FALSE);\n",
    "return(pruned)\n",
    "}\n",
    "\n",
    "# create function for KING-robust analysis\n",
    "king_mat=function(gds){\n",
    "samp.id=read.gdsn(index.gdsn(gds, \"sample.id\"))\n",
    "ibd.robust=snpgdsIBDKING(gds, sample.id=samp.id, family.id=NULL, maf=0.01,missing.rate=0.01,num.thread=cores)\n",
    "return(ibd.robust)\n",
    "}\n",
    "\n",
    "\n",
    "# create function for PCAiR\n",
    "pcair_r=function(gds_geno, pruned, KINGmat){\n",
    "pcair=pcair(gds_geno, snp.include=pruned,\n",
    "            kinobj=KINGmat, divobj=KINGmat)\n",
    "return(pcair)\n",
    "}\n",
    "\n",
    "# create function for PCAiR with sample filter\n",
    "pcair_r_fil=function(gds_geno, pruned, KINGmat, sample_list){\n",
    "    pcair=pcair(gds_geno, snp.include=pruned,\n",
    "                sample.include=sample_list,\n",
    "                kinobj=KINGmat, divobj=KINGmat)\n",
    "                return(pcair)\n",
    "                }\n",
    "\n",
    "\n",
    "# create function to match the sample ID in the GDS with the sample IDs of the supplied sampleID filter\n",
    "filter_samples=function(gds_samples, incl_samples){\n",
    "    matches=unique(grep(paste(incl_samples$SampleID,collapse=\"|\"),\n",
    "                        gds_samples, value=TRUE))\n",
    "    return(matches)\n",
    "}\n",
    "\n",
    "\n",
    "# run analysis for pcs with relationships\n",
    "# open the gds object\n",
    "gds=snpgdsOpen(opt$gdsfile);\n",
    "# LD prunning\n",
    "pruned=ld_prun(gds)\n",
    "# build KING matrix\n",
    "KINGmat=king_mat(gds)\n",
    "# adjust KING matrix\n",
    "KINGmat_m=KINGmat$kinship\n",
    "# add sampleIDs to colnames y row names\n",
    "colnames(KINGmat_m)=KINGmat$sample.id\n",
    "rownames(KINGmat_m)=KINGmat$sample.id\n",
    "# get samples in gds\n",
    "gds_samples=read.gdsn(index.gdsn(gds, \"sample.id\"))\n",
    "# close the gds object\n",
    "snpgdsClose(gds)\n",
    "\n",
    "# read the gds object\n",
    "gds=GdsGenotypeReader(filename=opt$gdsfile)\n",
    "\n",
    "# create a GenotypeData class object\n",
    "gds_geno=GenotypeData(gds)\n",
    "\n",
    "# run PCAiR\n",
    "PCair=pcair_r(gds_geno, pruned, KINGmat_m)\n",
    "PCairpart=pcairPartition(kinobj = KINGmat_m, divobj = KINGmat_m)\n",
    "\n",
    "# PC relate calculation \n",
    "gdsData=GenotypeBlockIterator(gds_geno, snpInclude=pruned)\n",
    "PCrelate=pcrelate(gdsData,\n",
    "                  pcs=PCair$vectors[,1:2],\n",
    "                  training.set=PCairpart$unrels,\n",
    "                  BPPARAM=BiocParallel::SerialParam())\n",
    "\n",
    "# making GRM\n",
    "grm=pcrelateToMatrix(PCrelate)\n",
    "# making sparce matrix\n",
    "grm[grm<0.05]=0\n",
    "grm_sparse=as.matrix(grm, sparse = TRUE)\n",
    "\n",
    "# save file\n",
    "out=list()\n",
    "out$PCair=PCair\n",
    "out$PCairpart=PCairpart\n",
    "out$KINGmat=KINGmat_m\n",
    "out$pruned=pruned\n",
    "out$PCrelate=PCrelate\n",
    "out$grm=grm\n",
    "out$grm_sparse=grm_sparse\n",
    "saveRDS(file=paste0(opt$out,\".rds\",\"\"),out)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "072ce1d8-71f4-4003-8daa-ffbaa6745556",
   "metadata": {},
   "source": [
    "### 3.2. Generar los archivos de PCA y GRM.\n",
    "***Descripción:***\n",
    "\n",
    "Este script ejecutará el \"create_grm.Rscript\" creado en el paso anterior.\n",
    "Deberá ajustar este paso según sus propios archivos de entrada.\n",
    "La salida será un archivo .RDS que contendrá tanto la información de PCA como la Matriz de Relaciones Genéticas (GRM)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e68639b-7e4b-41db-b674-25823fdcb838",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run R script for creating GDS files\n",
    "# add the path of the create_grm.Rscript from the previous step\n",
    "Rscript=\"create_grm.Rscript\"\n",
    "Rscript $Rscript \\\n",
    "--gdsfile=gdsfile_prunned \\\n",
    "--out=cohort_name.gds_prunned_grm_pca"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d9f7e11-1d7d-4881-bc6a-9a30a1d2368b",
   "metadata": {},
   "source": [
    "### 3.3. Guardar el archivo de PCA\n",
    "\n",
    "***Descripción:***\n",
    "Este script extrae los 10 principales componentes del archivo .RDS generado en el paso anterior y los guarda en un archivo separado para ser usados como covariables en Regenie."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "001a5026-f1c5-4b23-a249-b367101a4f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the PCA/GRM result file\n",
    "pca_grm <- readRDS(\"cohort_name.gds_prunned_grm_pca.rds\")\n",
    "\n",
    "# Extract sample IDs (IIDs)\n",
    "iids <- rownames(pca_grm$PCair$vectors)\n",
    "\n",
    "# Extract top 10 PCs\n",
    "pcs <- pca_grm$PCair$vectors[,c(1:10)]\n",
    "pcs <- as.data.frame(pcs)\n",
    "colnames(pcs) <- paste0(rep(\"PC\",10),rep(1:10))\n",
    "\n",
    "# add sample ID\n",
    "pcs$SampleID=rownames(pcs)\n",
    "\n",
    "covariate_table <- pcs\n",
    "\n",
    "# Save to file\n",
    "write.table(covariate_table,\n",
    "            file = \"cohort_10pcs_forregenie.txt\",\n",
    "            quote = FALSE,\n",
    "            row.names = FALSE,\n",
    "            col.names = TRUE,\n",
    "            sep = \"\\t\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
