{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ec5226fd",
   "metadata": {},
   "source": [
    "# Array Quality Control\n",
    "Authors: Rafaella Ormond and Jose Jaime Martinez-Magana <br>\n",
    "***Description:***<br>\n",
    "This script will prepare the files for perform for imputation\n",
    "\n",
    "**INPUT:** PLINK files (.bed/.bim/.fam)<br>\n",
    "**OUTPUT:** \n",
    "1) PCA results files (principal components table for samples).\n",
    "2) Genetic Relationship Matrix (GRM) files.\n",
    "\n",
    "### ***Requirements:***\n",
    "### Download Plink\n",
    "We can download Plink version 1.9 and version 2.0 following the steps from their website.<br>\n",
    "For instal plink2 [access here](https://www.cog-genomics.org/plink/2.0/)<br>\n",
    "For instal plink1.9 [access here](https://www.cog-genomics.org/plink/1.9/) <br>\n",
    "\n",
    "### Download BCF tools:\n",
    "For instal bcftools [acess here](http://samtools.github.io/bcftools/) <br>\n",
    "\n",
    "### Reference Genome: \n",
    "For the first steps the reference genome depends of the array, can be used hg19 or hg38<br>\n",
    "For download the reference genome hg19 please download the file: https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/hg19.fa.gz<br> \n",
    "For download the reference genome hg38 please download the file: https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz\n",
    "\n",
    "### Imputation on TOPMED\n",
    "After the execution of this script, the imputation needs to be performed on TOPMED [link here](https://imputation.biodatacatalyst.nhlbi.nih.gov/#!)<br>\n",
    "\n",
    "**Note:** After imputation, all files should be aligned to the **hg38** reference genome for consistency.\n",
    "\n",
    "### Analysis Steps:\n",
    "1. **User Configuration** – Define paths, reference genome, and software parameters.  \n",
    "2. **Convert PLINK to VCF** – Convert `.bed/.bim/.fam` files to `.vcf` format and index them.  \n",
    "3. **Alignment** – Align genotypes to the reference genome (strand checking).  \n",
    "4. **Split by Chromosome** – Generate one VCF file per chromosome (`chr1.vcf.gz`, `chr2.vcf.gz`, ...).\n",
    "5. **Instructions for TOPMED imputation** - Generate an imputated file in hg38 and phased"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "805326f2-dbc5-470e-8d78-b0a22b7f838d",
   "metadata": {},
   "source": [
    "### 1) User Configuration\n",
    "***Description:***<br>\n",
    "Set the parameters and adjust the path and files according to your analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "af55a9d8-1d98-4bae-8344-245aea311643",
   "metadata": {},
   "outputs": [],
   "source": [
    "# USER CONFIGURATION - MODIFY ACCORDING TO YOUR ANALYSIS\n",
    "\n",
    "# PLINK files prefix (.bed, .bim, .fam)\n",
    "inPath=\"/path_to_your_data/plink_prefix\"\n",
    "\n",
    "# Your cohort name\n",
    "cohortName=\"cohort_name\"\n",
    "\n",
    "# Directory for VCF files\n",
    "clvcfPath=\"path_vcf_files\"\n",
    "\n",
    "# Directory for VCFs by chromosome\n",
    "clvcfbychrPath=\"path_vcf_chr\"\n",
    "\n",
    "# Reference genome file (hg19 or hg38)\n",
    "refFile=\"/path/to/reference/hg19.fa\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1628e4d8-2937-4eaa-b5b1-c20f16ad3bfb",
   "metadata": {},
   "source": [
    "### 2) Convert PLINK files to VCF, perform quality control, and index the files\n",
    "***Description:***<br>\n",
    "Convert PLINK binary files to compressed VCF format and create index files for efficient access.  \n",
    "Quality control filters are applied during conversion to exclude variants and samples based on minor allele frequency (MAF), Hardy-Weinberg equilibrium (HWE), genotype missingness per variant and per individual. \n",
    "\n",
    "Quality control filters: <br>\n",
    "--maf 0.01  → exclude variants with minor allele frequency < 1%<br>\n",
    "--hwe 1e-6  → exclude variants with Hardy-Weinberg p-value < 1e-6<br>\n",
    "--geno 0.05 → exclude variants with missing genotype rate > 5%<br>\n",
    "--mind 0.05 → exclude samples with missing genotype rate > 5%<br>\n",
    "\n",
    "Please adjust the input paths and filenames accordingly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1acb1b9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert PLINK binary files (.bed/.bim/.fam) to a compressed VCF (.vcf.gz) and aply que quality control filters\n",
    "# - 'bgz' outputs a block-gzipped (bgzip) VCF file compatible with bcftools/tabix\n",
    "# The flag 'vcf-iid bgz' removes the Family ID (FID) field, keeping only the Individual ID (IID)\n",
    "plink2 --bfile ${inPath} --chr 1-22 --recode vcf-iid bgz --maf 0.01 --hwe 1e-6 --geno 0.05 --mind 0.05 --out ${clvcfPath}/${cohort_name}\n",
    "\n",
    "# Index the resulting VCF file using bcftools\n",
    "# This creates a .csi index file, needed for fast access and downstream analyses\n",
    "bcftools index ${clvcfPath}/${cohort_name}.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71d8c4a9-1f69-4ee3-9016-a52f560c367a",
   "metadata": {},
   "source": [
    "### 3) Aligment\n",
    "***Description:***<br>\n",
    "Align VCF alleles with the reference genome<br>\n",
    "Please adjust the threads accordingly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46c2d8fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert VCF file to BCF format using 25 threads for faster processing (adjusting acordly your compuationak power)\n",
    "bcftools convert ${clvcfPath}/${cohort_name}.vcf.gz -Ob -o ${clvcfPath}/${cohort_name}.bcf.gz --threads 25\n",
    "\n",
    "# Run BCFtools fixref plugin to fix reference/alternative allele inconsistencies\n",
    "# This step checks and updates alleles based on the reference genome\n",
    "bcftools +fixref ${clvcfPath}/${cohort_name}.bcf.gz -- -f ${refFile}\n",
    "\n",
    "# Run BCFtools fixref again, this time outputting a new VCF file\n",
    "# The -d option removes alleles that do not match the reference\n",
    "# The -m flip option flips alleles to match the reference when possible\n",
    "# The final result is a VCF file corrected for reference mismatches\n",
    "bcftools +fixref ${clvcfPath}/${cohort_name}.bcf.gz -Oz -o ${clvcfPath}/${cohort_name}.vcf.gz -- -d -m flip -f ${refFile}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9380d5b8-87cc-4ad4-8c71-eba4549b2e92",
   "metadata": {},
   "source": [
    "### 4) Split VCF file by chromosome\n",
    "***Description:***<br>\n",
    "Split VCF file by chromosome for imputation and sort VCF files by chromosome to ensure proper ordering before downstream analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b41e439",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate over chromosome numbers\n",
    "for chr_num in {1..22}\n",
    "do\n",
    "plink2 --vcf ${clvcfPath}/${cohort_name}.vcf.gz --chr ${chr_num} --recode vcf-4.2 bgz --out ${clvcfbychrPath}/${cohort_name}_forimputation_chr${chr_num}\n",
    "done\n",
    "\n",
    "# sorting\n",
    "# Iterate over chromosome numbers\n",
    "for chr_num in {1..22}\n",
    "do\n",
    "bcftools sort ${clvcfbychrPath}/${cohort_name}_forimputation_chr${chr_num}.vcf.gz -Oz -o ${clvcfbychrPath}/${cohort_name}_forimputation_chr${chr_num}_sorted.vcf.gz\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f5e25a5-e5a9-46b5-9473-aa2f96a4e6ca",
   "metadata": {},
   "source": [
    "## 5) Instructions for Imputation on TOPMed\n",
    "\n",
    "Please access the [TOPMed Imputation Server](https://imputation.biodatacatalyst.nhlbi.nih.gov).\n",
    "\n",
    "Log in to your account and access the dashboard.  \n",
    "At the top of the page, click **“Run” → “Genotype Imputation (Minimac4)”**.\n",
    "\n",
    "<p align=\"center\">\n",
    "  <img src=\"topmed1.png\" alt=\"Example of TOPMed interface\" width=\"600\">\n",
    "</p>\n",
    "\n",
    "Upload your VCF files and configure the following options:\n",
    "\n",
    "- **Array Build:** select *hg19* or *hg38* (according to your file build).  \n",
    "- **Rsq Filter:** *off*  \n",
    "- **Phasing:** *Eagle v2.4 (phased output)*  \n",
    "- **Population:** select *All populations* (currently “vs. TOPMed Panel”)  \n",
    "- **Mode:** *Quality Control & Imputation*  \n",
    "- Enable the checkbox **“Generate Meta-imputation file”**\n",
    "\n",
    "After submitting the job, you will receive an email when the results are ready.  \n",
    "You will also receive a **password by email** to access the output folders.  \n",
    "Download all the result folders to your working directory.\n",
    "\n",
    "<p align=\"center\">\n",
    "  <img src=\"topmed2.png\" alt=\"Example of download page\" width=\"600\">\n",
    "</p>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe38f500-12a2-4cf6-a5ac-18b0026a40c2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
