{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Post-Imputation Data Processing\n",
    "\n",
    "Authors: Rafaella Ormond and Jose Jaime Martinez-Magana\n",
    "\n",
    "***Description:***\n",
    "This script processes imputed data from TOPMED server, converting VCF files to PLINK format and performing quality control filtering.\n",
    "\n",
    "**INPUT:** Imputation output files from TOPMed server (chromosome-specific VCFs).\n",
    "**OUTPUT:** Quality-controlled PLINK dataset files, filtered and ready for downstream analyses.\n",
    "\n",
    "### Requirements:\n",
    "- PLINK 1.9 or 2.0 ([plink 2.0](https://www.cog-genomics.org/plink/2.0/), [plink 1.9](https://www.cog-genomics.org/plink/1.9/))\n",
    "- BCFtools ([BCFtools Download](http://samtools.github.io/bcftools/))\n",
    "- TOPMED Imputation Results (VCF files per chromosome, QC report, summary stats)\n",
    "\n",
    "### Recommended Workflow:\n",
    "1. Unzip imputation output files\n",
    "2. PLINK: create and merge chromosome-specific files\n",
    "3. BCFtools: QC, convert to PLINK, merge, and standardize variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Unzip the imputation files\n",
    "***Description:***\n",
    "Unzip the output files from imputation. Some servers encrypt the files with a password."
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "execution_count": null,
   "source": [
    "# Set path to imputation files\n",
    "input_path=\"/path_to_your_data/imputation_results\"\n",
    "password='YOUR_PASSWORD'\n",
    "\n",
    "cd ${input_path}\n",
    "for chr in *.zip; do\n",
    "    unzip -P ${password} -o ${chr}\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. PLINK\n",
    "Convert the `.dose.vcf.gz` files into PLINK format and merge all chromosomes."
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "execution_count": null,
   "source": [
    "# User configuration\n",
    "plink2=\"/path_to_your_data/plink2\"\n",
    "splitted_path=\"/path_to_your_data/splitted_files\"\n",
    "merged_path=\"/path_to_your_data/merged_files\"\n",
    "ldpruned_path=\"/path_to_your_data/merged_files_ldpruned\"\n",
    "input_vcf_path=\"/path_to_your_data/imputation_results\""
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "execution_count": null,
   "source": [
    "# Transform VCF dose files to PLINK format by chromosome\n",
    "for chr in {1..22}; do\n",
    "    ${plink2} --vcf ${input_vcf_path}/chr${chr}.dose.vcf.gz \\\n",
    "        --make-pgen --id-delim '_' --rm-dup 'force-first' --snps-only \\\n",
    "        --out ${splitted_path}/chr${chr}.dose.plink\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge PLINK files and standardize variant IDs"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "execution_count": null,
   "source": [
    "cd ${splitted_path}\n",
    "# Create list of chromosome files (exclude chr1, will be base)\n",
    "for chr in {2..22}; do\n",
    "    echo chr${chr}.dose.plink >> plink_files.list\n",
    "done\n",
    "\n",
    "# Merge all chromosomes into single dataset\n",
    "${plink2} --pfile chr1.dose.plink --pmerge-list plink_files.list pfile --multiallelics-already-joined --out ${merged_path}/merged.chr.dose.plink\n",
    "\n",
    "# Standardize variant IDs to chr:bp:ref:alt and keep only SNPs\n",
    "${plink2} --pfile ${merged_path}/merged.chr.dose.plink --set-all-var-ids 'chr@:#:$1:$2' --snps-only 'just-acgt' --make-pgen --out ${merged_path}/merged.chr.dose.plink.snpsonly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. BCFtools"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "execution_count": null,
   "source": [
    "# User configuration\n",
    "input='/path_to_your_data/00impfiles'\n",
    "filtered_path='/path_to_your_data/02impfilesfiltered'\n",
    "plink_out='/path_to_your_data/plink_per_chr'\n",
    "merged_path='/path_to_your_data/merged_plink'\n",
    "plink2='/path_to_your_data/plink2'"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "execution_count": null,
   "source": [
    "# Quality control and filtering using BCFtools\n",
    "for chr in {1..22}; do\n",
    "    bcftools view -i 'R2>0.8 & MAF>0.001' --threads 30 -Oz -o ${filtered_path}/cohort.chr${chr}.dose.filtered.vcf.gz ${input}/chr${chr}.dose.vcf.gz\n",
    "    bcftools +fill-tags --threads 30 ${filtered_path}/cohort.chr${chr}.dose.filtered.vcf.gz --write-index -Ob -o ${filtered_path}/cohort.chr${chr}.dose.filtered.tag.bcf.gz -- -t AC,AN\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "execution_count": null,
   "source": [
    "# Convert filtered BCF files to PLINK format\n",
    "for chr in {1..22}; do\n",
    "    ${plink2} \\\n",
    "      --bcf ${filtered_path}/cohort.chr${chr}.dose.filtered.tag.bcf.gz \\\n",
    "      --max-alleles 2 \\\n",
    "      --set-all-var-ids 'chr@:#:$1:$2' \\\n",
    "      --new-id-max-allele-len 100 \\\n",
    "      --make-pgen \\\n",
    "      --out ${plink_out}/cohort_chr${chr}.plink\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "execution_count": null,
   "source": [
    "# Merge chromosome-specific PLINK files\n",
    "mkdir -p ${merged_path}\n",
    "\n",
    "# Create merge list (exclude chr1)\n",
    "for chr in {2..22}; do\n",
    "    echo \"${plink_out}/cohort_chr${chr}.plink\" >> ${merged_path}/cohort_merge_list.txt\n",
    "done\n",
    "\n",
    "# Remove duplicates and merge\n",
    "for chr in {1..22}; do\n",
    "    ${plink2} --pfile ${plink_out}/cohort_chr${chr}.plink --rm-dup 'exclude-all' --make-pgen --out ${merged_path}/cohort_nodup_chr${chr}\n",
    "done\n",
    "\n",
    "# Merge all non-duplicate chromosome files (chr1 as base)\n",
    "cd ${merged_path}\n",
    "${plink2} --pfile cohort_nodup_chr1 --pmerge-list cohort_merge_list.txt 'pfile' --make-pgen --out cohort_nodup_merged"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "execution_count": null,
   "source": [
    "# Standardize variant IDs and keep only SNPs\n",
    "${plink2} \\\n",
    "  --pfile ${merged_path}/cohort_nodup_merged \\\n",
    "  --set-all-var-ids 'chr@:#:$1:$2' \\\n",
    "  --snps-only 'just-acgt' \\\n",
    "  --make-pgen \\\n",
    "  --out ${merged_path}/cohort_nodup_merged.snpsonly"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "name": "sos"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
