{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b96c4124-200e-4010-8537-391759951dc3",
   "metadata": {},
   "source": [
    "# Post-Imputation Data Processing\n",
    "\n",
    "Authors: Rafaella Ormond and Jose Jaime Martinez-Magana <br>\n",
    "***Description:***<br>\n",
    "This script processes imputed data from TOPMED server, converting VCF files to PLINK format and performing quality control filtering\n",
    "\n",
    "**INPUT:** Imputation output files from TOPMed server (e.g., chromosome-specific VCFs or dosage files).<br>\n",
    "**OUTPUT:** Quality-controlled PLINK dataset files, filtered and ready for downstream analyses.\n",
    "\n",
    "\n",
    "### ***Requirements:***\n",
    "### Download Plink\n",
    "You can download Plink version 1.9 and version 2.0 following the steps from their website.<br>\n",
    "For install plink2 [access here](https://www.cog-genomics.org/plink/2.0/)<br>\n",
    "For install plink1.9 [access here](https://www.cog-genomics.org/plink/1.9/) \n",
    "\n",
    "### Download BCFtools\n",
    "**BCFtools** is another popular tool for handling VCF/BCF files:  \n",
    "- [BCFtools Download](http://samtools.github.io/bcftools/)  \n",
    "\n",
    "### TOPMED Imputation Results\n",
    "This script assumes you have already completed imputation using TOPMED server [link here](https://imputation.biodatacatalyst.nhlbi.nih.gov/#!) and downloaded the results<br>\n",
    "After imputation is complete, you will need to download:\n",
    "\n",
    "- The **Quality Control Report** (e.g., `qcreport.html`)\n",
    "- The **QC summary statistics** (e.g., `chunks-excluded.txt`, `snps-excluded.txt`, `typed-only.txt`)\n",
    "- The **imputed VCF files**, one per chromosome (e.g., `chr1.dose.vcf.gz`, `chr2.dose.vcf.gz`, ...)\n",
    "\n",
    "These files are available via direct links from the server interface after your job completes. Make sure to save them in a consistent folder structure for downstream analysis.\n",
    "\n",
    "## QC Software Options\n",
    "\n",
    "You can perform post-imputation quality control using one of the following:\n",
    "\n",
    "1) Plink\n",
    "2) BCFtools\n",
    "\n",
    "| Tool      | Pros ‚úÖ | Cons ‚ùå |\n",
    "|-----------|--------|---------|\n",
    "| **PLINK** | - Widely used in GWAS pipelines.<br>- Very fast for large datasets.<br>- Many built-in genetic QC functions (MAF, HWE, missingness, etc.).<br>- Can output multiple formats. | - Requires conversion from VCF to PLINK format first.<br>- Limited manipulation of complex VCF structures compared to BCFtools. |\n",
    "| **BCFtools** | - Works directly with compressed VCF/BCF files.<br>- Very efficient for filtering and subsetting large variant files.<br>- Flexible for script automation.<br>- Supports streaming and piping with other tools. | - Requires more manual scripting for GWAS-specific QC steps.<br>- Less \"turnkey\" for genetic association workflows compared to PLINK. |\n",
    "\n",
    "---\n",
    "\n",
    "üí° **Tip:**  \n",
    "- If you plan to move quickly into GWAS analysis, **PLINK** is usually the easiest choice.  \n",
    "- If you want maximum flexibility and work a lot with raw VCF files, **BCFtools** is a powerful option.\n",
    "\n",
    "Analysis Steps:¬∂\n",
    "1. **Unzip the imputation files**<br>\n",
    "2. **Plink**<br>\n",
    "   2.1 User Configuration<br>\n",
    "   2.2 Create plink files and merge all chromosomes<br>\n",
    "   2.3 Merge PLINK files and standardize variant IDs<br>\n",
    "   2.4 Perform LD pruning on merged PLINK dataset\n",
    "   \n",
    "3. **BCFTools**<br>\n",
    "   3.1 User Configuration<br>\n",
    "   3.2 Quality Control<br>\n",
    "   3.3 Convert Filtered BCF to PLINK by Chromosome<br>\n",
    "   3.4 Merge Chromosome-specific PLINK Files<br>\n",
    "   3.5 Standardize Variant IDs and Keep Only SNPs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fb5b9e3-5b57-4bc7-8425-0574c832a9fe",
   "metadata": {},
   "source": [
    "### 1. Unzip the imputation files\n",
    "***Description:***<br>\n",
    "Unzip the output files from imputation\n",
    "\n",
    "> **Note:** Some imputation servers, such as TOPMed, may encrypt the output files with a password for security purposes.  \n",
    "> The password is usually provided in your job results page or in the confirmation email.  \n",
    "> Make sure to replace the example password above with the one specific to your dataset.\n",
    "\n",
    "Please adjust the path and files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39f2a73f-a136-4629-a0a7-0f88d2169aab",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd ${input_path}\n",
    "\n",
    "# unzip all files\n",
    "# add the password of your work in the next line\n",
    "password='XXX'\n",
    "for chr in *.zip\n",
    "do\n",
    "    echo unzip -P ${password} -o ${chr}\n",
    "    unzip -P ${password} -o ${chr}\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9a32c31-487e-4b7a-b693-0c9f9ee0a18a",
   "metadata": {},
   "source": [
    "### 2. PLINK\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1cca72d-703d-4b9e-b67b-f608fe0fe2ea",
   "metadata": {},
   "source": [
    "### 2.1 User Configuration\n",
    "\n",
    "***Description:***<br>\n",
    "Set the parameters and adjust the path and files according to your analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e86e243c-e5b4-4170-8c80-8458f7bdcc06",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input path\n",
    "input_path=\"/path_to_your_data/imputation_results\"\n",
    "\n",
    "# set directory for files split by chromosome\n",
    "splitted_path=\"/path_to_your_data/splitted_files\"\n",
    "\n",
    "# set the plink2 file\n",
    "plink2=\"/path_to_your_data/plink2\"\n",
    "\n",
    "# set directory for files with the chromosomes merged\n",
    "merged_path=\"/path_to_your_data/merged_files\"\n",
    "\n",
    "# set ld pruned path\n",
    "ldpruned_path=\"/path_to_your_data/merged_files_ldpruned\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0893595c-b575-4f4b-b383-ffa0e20b8b76",
   "metadata": {},
   "source": [
    "### 2.2 Create plink files and merge all chromosomes\n",
    "***Description:***<br>\n",
    "Convert the `.dose.vcf.gz` files into PLINK format and merge all chromosomes into a single dataset. \n",
    "\n",
    "Please adjust the path and files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35f1c310-c87f-4fb3-a509-b3b32b1ba275",
   "metadata": {},
   "outputs": [],
   "source": [
    "# transform dose files to plink\n",
    "for chr in {1..22}\n",
    "do\n",
    "${plink2} --vcf ${input_path}/chr${chr}.dose.vcf.gz --make-pgen --id-delim '_' --rm-dup 'force-first' --snps-only --out ${splitted_path}/chr${chr}.dose.plink\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c9334e5-fc6e-4586-ac27-682ce7bf04ba",
   "metadata": {},
   "source": [
    "### 2.3 Merge PLINK files and standardize variant IDs\n",
    "\n",
    "***Description:*** <br>\n",
    "This script moves to the directory containing PLINK files split by chromosome, creates a list of those files, merges them into a single dataset, and standardizes variant IDs to the format chr:bp:ref:alt for consistency in downstream analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57d80c10-bbc4-4a84-842c-7ba4d5395e71",
   "metadata": {},
   "outputs": [],
   "source": [
    "# move to splitted by chromosome path\n",
    "cd ${splitted_path}\n",
    "\n",
    "# create list of plink files if needed\n",
    "# for chr in {2..22};do echo chr${chr}.dose.plink;done > plink_files.list\n",
    "\n",
    "# load libraries\n",
    "plink2=\"/path_to_your_data/plink2\"\n",
    "\n",
    "# transform dose files to plink\n",
    "${plink2} --pfile chr1.dose.plink --pmerge-list plink_files.list pfile --multiallelics-already-joined --out ${merged_path}/merged.chr.dose.plink\n",
    "\n",
    "# change all variants ID to chr:bp:ref:alt\n",
    "${plink2} --pfile ${merged_path}/merged.chr.dose.plink --set-all-var-ids 'chr@:#:$1:$2' --snps-only 'just-acgt' --make-pgen --out ${merged_path}/merged.chr.dose.plink.norsids.snpsonly"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56c8e639-31f9-459c-9791-294ead4461d9",
   "metadata": {},
   "source": [
    "### 3. BCFTools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "437440b2-a681-470b-90ba-32e04912e9a2",
   "metadata": {},
   "source": [
    "### 3.1 User Configuration\n",
    "\n",
    "***Description:***<br>\n",
    "Set the parameters and adjust the path and files according to your analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8b1bd17-7a22-4e58-a310-a8649d104393",
   "metadata": {},
   "outputs": [],
   "source": [
    "input='/path_to_your_data/00impfiles'\n",
    "filtered_path='/path_to_your_data/02impfilesfiltered'\n",
    "plink_path=\"/path_to_your_data/plink2\"\n",
    "plink_out=\"/path_to_your_data/plink_per_chr\"\n",
    "merged_path=\"/path_to_your_data/merged_plink\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05301a4c-6d94-49be-93dd-424dff4bc1eb",
   "metadata": {},
   "source": [
    "### 3.2 Quality Control\n",
    "\n",
    "**Description:**  \n",
    "This script sequentially processes chromosome-specific imputed variant files by applying quality filters using `bcftools`. Variants are filtered based on imputation quality (R2) and minor allele frequency (MAF). The filtered files are then compressed and indexed, preparing them for downstream genetic analyses.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b5e6f32-5456-45db-81f7-c984f0f3cb32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter BRHC imputed variants\n",
    "# set the paths for input and output data\n",
    "\n",
    "# loop over each chromosome sequentially\n",
    "for chr in {1..22}; do\n",
    "    bcftools view -i 'R2>0.8 & MAF>0.001' --threads 30 -Oz -o ${filtered_path}/cohort.chr${chr}.dose.filtered.vcf.gz ${input}/chr${chr}.dose.vcf.gz\n",
    "    bcftools +fill-tags --threads 30 ${filtered_path}/cohort.chr${chr}.dose.filtered.vcf.gz --write-index -Ob -o ${filtered_path}/cohort.chr${chr}.dose.filtered.tag.bcf.gz -- -t AC,AN\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7331e2a-c1f0-432f-bbe3-53b4c3242d67",
   "metadata": {},
   "source": [
    "### 3.3 Convert Filtered BCF to PLINK by Chromosome\n",
    "***Description:***<br> \n",
    "Converts filtered BCF files to .pgen format using PLINK.<br>\n",
    "Change \"cohort\" to the file name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13eecaf8-5569-45f0-9c49-a9122f45517f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for chr in {1..22}; do\n",
    "plink2 \\\n",
    "  --bcf ${filtered_path}/cohort.chr${chr}.dose.filtered.tag.bcf.gz \\\n",
    "  --max-alleles 2 \\\n",
    "  --set-all-var-ids 'chr@:#:$1:$2' \\\n",
    "  --new-id-max-allele-len 100 \\\n",
    "  --make-pgen \\\n",
    "  --out ${plink_out}/cohort_chr${chr}.plink\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63acee2c-7e38-4e7c-82ca-d17e08324854",
   "metadata": {},
   "source": [
    "### 3.4 Merge Chromosome-specific PLINK Files\n",
    "***Description:*** \n",
    "Merges individual chromosome `.pgen` files into a single PLINK dataset.\n",
    "\n",
    "Please modify the names acordly with your file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfc684df-be0c-474b-bb41-4f2a309540ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# move to de path\n",
    "mkdir -p ${merged_path}\n",
    "\n",
    "# Creat list of files to merge (exclude the chromosome 1, will be the base file)\n",
    "for chr in {2..22}; do\n",
    "    echo \"${plink_out}/cohort_chr${chr}.plink\" >> ${merged_path}/cohort_merge_list.txt\n",
    "done\n",
    "\n",
    "# Removing all duplicates\n",
    "for chr in {1..22}\n",
    "do\n",
    "plink2 --pfile ${datapath}/cohort_chr${chr}.plink --rm-dup 'exclude-all' --make-pgen --out cohort_nodup_chr${chr} \\\n",
    "done\n",
    "\n",
    "# Getting the file for merging\n",
    "current_path= your_path\n",
    "for chr in {2..22}\n",
    "do\n",
    "echo \"${current_path}/cohort_nodup_chr${chr}\" >> cohort_merging-09182025.txt\n",
    "done\n",
    "\n",
    "# Merging\n",
    "plink2 --pfile cohort_nodup_chr1 --pmerge-list cohort_merging.txt 'pfile' --make-pgen --out cohort_nodup_merged"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc83af6d-3a22-476e-a49a-eba1f3e1c207",
   "metadata": {},
   "source": [
    "### 3.5 Standardize Variant IDs and Keep Only SNPs\n",
    "***Description:*** <br>\n",
    "Updates variant IDs to the `@:#:$r:$a` format, filters to include only A/C/G/T SNPs, and recreates the `.pgen` file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6afd62d-5464-4012-9dfc-8141fce2ab87",
   "metadata": {},
   "outputs": [],
   "source": [
    "${plink_path} \\\n",
    "    --pfile ${merged_path}/cohort_merged_allchr \\\n",
    "    --set-all-var-ids \\\n",
    "    --snps-only just-acgt \\\n",
    "    --make-pgen \\\n",
    "    --out ${merged_path}/cohort_merged_allchr.snpsonly"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
