{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ec5226fd",
   "metadata": {},
   "source": [
    "# Array QC\n",
    "Author: Rafaella Ormond and Jose Jaime Martinez-Magana <br>\n",
    "Description:<br>\n",
    "This script will prepare the files for perform the GWAS of smoking traits on LAGC cohorts\n",
    "\n",
    "### ***Requirements:***\n",
    "### Download Plink\n",
    "We can download Plink version 1.9 and version 2.0 following the steps from their website.<br>\n",
    "For instal plink2 [access here](https://www.cog-genomics.org/plink/2.0/)<br>\n",
    "For instal plink1.9 [access here](https://www.cog-genomics.org/plink/1.9/) <br>\n",
    "\n",
    "### Download BCF tools:\n",
    "For instal bcftools [acess here](https://www.htslib.org/download/) <br>\n",
    "\n",
    "### Reference Genome: \n",
    "For the first steps the reference genome depends of the array, can be used hg19 or hg38<br>\n",
    "For download the reference genome hg19 please download the file: https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/hg19.fa.gz<br> \n",
    "For download the reference genome hg38 please download the file: https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz\n",
    "\n",
    "### Imputation on TOPMED\n",
    "After the execution of this script, the imputation needs to be performed on TOPMED [link here](https://imputation.biodatacatalyst.nhlbi.nih.gov/#!)<br>\n",
    "After the imputation all files needs to be transformed to hg38"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff39bb79-cf7b-463d-b73a-75433f9b0c07",
   "metadata": {},
   "source": [
    "Ask for computational resources if needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1999b822",
   "metadata": {},
   "outputs": [],
   "source": [
    "# request computational resources if needed\n",
    "salloc -t 40:00 --mem=8G --partition=day"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "534fda3b-fe5f-4d62-8af8-9a1d1d09c074",
   "metadata": {},
   "source": [
    "Activate conda environment if needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ff9e535",
   "metadata": {},
   "outputs": [],
   "source": [
    "# activate environment\n",
    "module load miniconda\n",
    "conda activate step1_array_qc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "918e9c68-e157-43d7-bde6-7723dade138f",
   "metadata": {},
   "source": [
    "On this next step please change for your paths and files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a16c18d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## set parameters\n",
    "# imput path\n",
    "inPath= input_path\n",
    "# cohort path with vcf files\n",
    "clvcfPath='cohort_name_vcfs'\n",
    "# cohort path with vcf files slpited by chromosome\n",
    "clvcfbychrPath='cohort_name_vcfs_bychr'\n",
    "# Reference genmonic file to alignment\n",
    "refFile='hg19.fa.nochr.fa'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "574bc244",
   "metadata": {},
   "source": [
    "### Data for imputation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba804291-57a7-4ffb-b7a7-31ead7cde8d7",
   "metadata": {},
   "source": [
    "Convert PLINK files to VCF and index the output<br>\n",
    "Please adjust the path and files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1acb1b9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PLINK create vcf command\n",
    "# this flag 'vcf-iid bgz' will remove the FID\n",
    "plink --bfile ${inPath} --recode vcf-iid bgz --maf 0.01 --hwe 1e-6 --geno 0.05 --mind 0.05 --out ${clvcfPath}/cohort_name\n",
    "\n",
    "#index\n",
    "bcftools index ${clvcfPath}/cohort_name.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71d8c4a9-1f69-4ee3-9016-a52f560c367a",
   "metadata": {},
   "source": [
    "Align VCF alleles with the reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "46c2d8fd",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (1077626377.py, line 2)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[1], line 2\u001b[0;36m\u001b[0m\n\u001b[0;31m    bcftools convert ${clvcfPath}/cohort_name.vcf.gz -Ob -o ${clvcfPath}/cohort_name.bcf.gz --threads 25\u001b[0m\n\u001b[0m             ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "# Convert VCF file to BCF format using 25 threads for faster processing\n",
    "bcftools convert ${clvcfPath}/cohort_name.vcf.gz -Ob -o ${clvcfPath}/cohort_name.bcf.gz --threads 25\n",
    "\n",
    "# Run BCFtools fixref plugin to fix reference/alternative allele inconsistencies\n",
    "# This step checks and updates alleles based on the reference genome\n",
    "bcftools +fixref ${clvcfPath}/cohort_name.bcf.gz -- -f ${refFile}\n",
    "\n",
    "# Run BCFtools fixref again, this time outputting a new VCF file\n",
    "# The -d option removes alleles that do not match the reference\n",
    "# The -m flip option flips alleles to match the reference when possible\n",
    "# The final result is a VCF file corrected for reference mismatches\n",
    "bcftools +fixref ${clvcfPath}/cohort_name.bcf.gz -Oz -o ${clvcfPath}/cohort_name.vcf.gz -- -d -m flip -f ${refFile}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9380d5b8-87cc-4ad4-8c71-eba4549b2e92",
   "metadata": {},
   "source": [
    "Split VCF file by chromosome for imputation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b41e439",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate over chromosome numbers\n",
    "for chr_num in {1..22}\n",
    "do\n",
    "plink2 --vcf ${clvcfPath}/cohort_name.vcf.gz --chr ${chr_num} --recode vcf-4.2 bgz --out ${clvcfbychrPath}/cohort_name_forimputation_chr${chr_num}\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0113731b-61f9-471f-aaf5-c8a135f864b8",
   "metadata": {},
   "source": [
    "Sort VCF files by chromosome to ensure proper ordering before downstream analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81b56b3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sorting\n",
    "# Iterate over chromosome numbers\n",
    "for chr_num in {1..22}\n",
    "do\n",
    "bcftools sort ${clvcfbychrPath}/cohort_name_forimputation_chr${chr_num}.vcf.gz -Oz -o ${clvcfbychrPath}/cohort_name_forimputation_chr${chr_num}_sorted.vcf.gz\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
