{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6202ce00-100b-4ea0-8e0a-ef88bca29f6a",
   "metadata": {},
   "source": [
    "# GWAS Analysis on LAGC Cohorts Using Regenie\n",
    "Authors: Rafaella Ormond and Jose Jaime Martinez-Magana <br>\n",
    "\n",
    "***Description:***<br>\n",
    "In this analysis, a GWAS was performed using **Regenie** for one cohort in the LAGC.\n",
    "\n",
    "**INPUT:**\n",
    "1) Genotype data in PLINK pgen/pvar/sample format (.pgen, .pvar, .sample)\n",
    "2) Phenotype and covariate files corresponding to the cohort\n",
    "3) Covariates for adjustment in the GWAS\n",
    "\n",
    "**OUTPUT:** Regenie step 2 GWAS association results (regenie files with summary statistics)\n",
    "\n",
    "\n",
    "### ***Requirements:***\n",
    "\n",
    "**Regenie** is designed for use with cohorts that have a sample size greater than 300.\n",
    "\n",
    "### Download Plink\n",
    "We can download Plink version 1.9 and version 2.0 following the steps from their website.<br>\n",
    "For instal plink2 [access here](https://www.cog-genomics.org/plink/2.0/)<br>\n",
    "For instal plink1.9 [access here](https://www.cog-genomics.org/plink/1.9/) <br>\n",
    "\n",
    "### Download Regenie\n",
    "To download **Regenie**, follow this link: [Access here](https://github.com/rgcgithub/regenie)\n",
    "\n",
    "For more details, please refer to the analysis plan: [Access here](https://docs.google.com/document/d/1RzD5kBlj9rfiomda1G3NfxYDXLdmIUO7VX0cSNj70Kk/edit?usp=sharing)\n",
    "\n",
    "### Analysis Steps:\n",
    "1) Recode pfile to bfile<br>\n",
    "2) Recode pgen to bgen<br>\n",
    "3) Regenie step 1 (whole-genome regression model fitting)<br>\n",
    "4) Regenie step 2 (Association testing)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "493ca5ed-f62f-4d1b-a999-7a66c226bdf0",
   "metadata": {},
   "source": [
    "### 1) Recode pfile to bfile\n",
    "**Description:**<br>\n",
    "Recoding PLINK2 format files to bfiles format, if necessary<br>\n",
    "Please adjust the input paths and filenames accordingly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a13d4d1f-274e-48a1-86d4-7d38bf3712bc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## Transform pgen to bfile for Step 1\n",
    "# This step converts the .pgen format to .bfile format for use in step 1\n",
    "# Make sure to adjust the file names and suffixes according to your PLINK files\n",
    "\n",
    "plink2 \\\n",
    "    --pfile cohort_name \\\n",
    "    --maf 0.01 --mac 100 --geno 0.1 --hwe 1e-10 --mind 0.1 \\\n",
    "    --write-snplist --write-samples --no-id-header --make-bed --snps-only 'just-acgt' --max-alleles 2 \\\n",
    "    --out cohort_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8072c429-ed11-4ec3-9045-04838832f6bc",
   "metadata": {},
   "source": [
    "### 2) Recode pgen to bgen\n",
    "**Description:**<br>\n",
    "This step recodes PLINK2 format files (`.pgen`) to (`.bgen`) format for use in step 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9370174f-dffb-4a93-bee7-e7b08f4d6066",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Recode pgen to bgen \n",
    "# The pgen file gave an error in the psam file\n",
    "# Make sure to adjust the file names and suffixes according to your PLINK files\n",
    "\n",
    "plink2 --pfile cohort_name --export bgen-1.1 --snps-only 'just-acgt' --max-alleles 2 --out cohort_name --threads 30"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d60050b-f1e1-4651-8946-fd21c4a2f119",
   "metadata": {},
   "source": [
    "# Regenie\n",
    "\n",
    "For steps 1 and 2, phenotype and covariate files will be used, both separated by sex.  \n",
    "\n",
    "**Phenotype files (`inph`)** must include the following columns:  \n",
    "`FID`, `IID`, `Pheno1`, `Pheno2`  \n",
    "\n",
    "There should be two separate files:  \n",
    "- `inph_bt`: for **binary traits**  \n",
    "- `inph_qt`: for **quantitative traits**  \n",
    "\n",
    "The phenotypes must be selected and adjusted based on those available in your cohort and according to the analysis plan:  \n",
    "[Access the analysis plan here](https://docs.google.com/document/d/1RzD5kBlj9rfiomda1G3NfxYDXLdmIUO7VX0cSNj70Kk/edit?usp=sharing)  \n",
    "\n",
    "**Covariate file (`inco`)** must include the following columns:  \n",
    "`FID`, `IID`, `age`, and `PC1` to `PC10`  \n",
    "\n",
    "The 10 PCs (Principal Components) must be generated using **PC-AiR**.  \n",
    "Make sure the individuals in the phenotype and covariate files match.\n",
    "\n",
    "> **Note:** !!! Warning, we will use HapMap SNPs for Step 1!!!<br>\n",
    "> For hapmap we selected 250k snps randomly, please download this file on the github [LINK here](https://github.com/ormondr/Smoking_GWAS_LAGC/blob/main/English/02GWAS/00Regenie/w_hm3_hg38_random250K.snplist)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f1bab6b-e236-472d-9817-ff59f9e54c93",
   "metadata": {},
   "source": [
    "### 3) Regenie Step1\n",
    "\n",
    "**Description:**<br>\n",
    "**Step 1 – Whole-genome regression model fitting**<br>\n",
    "This step fits a ridge regression model across the genome using genotype data.  \n",
    "The goal is to estimate individual trait predictions (leave-one-chromosome-out, LOCO) while accounting for population structure and relatedness.  \n",
    "Input: genotype data in PLINK format and phenotype/covariate files.  \n",
    "Output: LOCO predictions used as covariates in Step 2."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54a2e51c-843d-4ce6-ac62-b870801b5203",
   "metadata": {},
   "source": [
    "> **Note:** !!! Warning, we will use HapMap SNPs for Step 1!!!<br>\n",
    "> For hapmap we selected 250k snps randomly, please download this file on the github [LINK here]()\n",
    "\n",
    "> **Note:** <br>\n",
    "> The `\"inph\"` files are the phenotype information<br>\n",
    "> The `\"inco\"` files are the covariate information, `\"bt\"` for binary traits and `\"qt\"` for quantitative traits<br>\n",
    "> `\"bt\"` for binary traits and \"qt\" for quantitative traits<br>\n",
    "> **The analysis needs to be separated by sex**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48869c60-92df-4817-9b19-29c073e40c85",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## Set parameters\n",
    "# Set genotype data\n",
    "# Adjust acording to your data\n",
    "# The \"inge\" needs to be the plink sufix files\n",
    "inge=\"path/cohort_name\"\n",
    "\n",
    "# The \"inph\" files are the phenotype information, \"bt\" for binary traits and \"qt\" for quantitative traits, separated by sex\n",
    "# Set phenotype and covariates files for females\n",
    "female_inph_bt=\"path/cohort_female_pheno_bt_forregenie.txt\"\n",
    "female_inph_qt=\"path/cohort_female_pheno_qt_forregenie.txt\"\n",
    "female_inco=\"path/cohort_female_covar_forregenie.txt\"\n",
    "\n",
    "# Set phenotype and covariates files for males\n",
    "male_inph_bt=\"path/cohort_male_pheno_bt_forregenie.txt\"\n",
    "male_inph_qt=\"path/cohort_male_pheno_qt_forregenie.txt\"\n",
    "male_inco=\"path/cohort_male_covar_forregenie.txt\"\n",
    "\n",
    "# Set output for step 1\n",
    "out_step1_qt_female=\"path/cohort/out_step1_qt_female\"\n",
    "out_step1_bt_female=\"path/cohort/out_step1_bt_female\"\n",
    "out_step1_qt_male=\"path/cohort/out_step1_qt_male\"\n",
    "out_step1_bt_male=\"path/cohort/out_step1_bt_male\"\n",
    "\n",
    "# Set output for step 2\n",
    "out_step2_qt_female=\"path/cohort/out_step2_qt_female\"\n",
    "out_step2_bt_female=\"path/cohort/out_step2_bt_female\"\n",
    "out_step2_qt_male=\"path/cohort/out_step2_qt_male\"\n",
    "out_step2_bt_male=\"path/cohort/out_step2_bt_male\"\n",
    "\n",
    "## Add zeros to the FID if you get errors. You could use the following code in bash\n",
    "# awk 'BEGIN {OFS=\"\\t\"} NR==1 {print \"FID\", $0; next} {print \"0\", $0}' \"${female_inph_qt}\" > tmpfile && mv tmpfile \"${female_inph_qt}\"\n",
    "## !!! Warning, we will use HapMap SNPs for Step 1!!!\n",
    "# For hapmap we selected 250k snps randomly, please download this file on the github\n",
    "hapmap=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/02lagc_smoking_gwas/11references/w_hm3_hg38_random250K.snplist\"\n",
    "\n",
    "## Running Step1\n",
    "# Running for females for quantitative traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${female_inco}\\\n",
    "    --phenoFile ${female_inph_qt}\\\n",
    "    --bsize 400\\\n",
    "    --qt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_qt_female}\n",
    "\n",
    "# Running for females for binary traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${female_inco}\\\n",
    "    --phenoFile ${female_inph_bt}\\\n",
    "    --bsize 400\\\n",
    "    --iid-only\\\n",
    "    --bt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_bt_female}\n",
    "\n",
    "# Running for males for quantitative traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${male_inco}\\\n",
    "    --phenoFile ${male_inph_qt}\\\n",
    "    --bsize 400\\\n",
    "    --iid-only\\\n",
    "    --qt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_qt_male}\n",
    "\n",
    "# Running for males for binary traits\n",
    "regenie\\\n",
    "    --step 1\\\n",
    "    --bed ${inge}\\\n",
    "    --covarFile ${male_inco}\\\n",
    "    --phenoFile ${male_inph_bt}\\\n",
    "    --bsize 400\\\n",
    "    --iid-only\\\n",
    "    --bt\\\n",
    "    --extract ${hapmap}\\\n",
    "    --force-step1\\\n",
    "    --out ${out_step1_bt_male}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ed2da9d-26d5-458a-8d28-e3951cdd5fa1",
   "metadata": {},
   "source": [
    "### 4) Regenie Step 2\n",
    "\n",
    "**Description:**\n",
    "**Step 2 – Association testing:** <br>\n",
    "This step performs the GWAS using the LOCO predictions from Step 1 as offsets in a linear/firth regression model.  \n",
    "It tests each variant for association with the trait of interest.  \n",
    "Input: same phenotype/covariate files, LOCO predictions, and full genotype data.  \n",
    "Output: GWAS results (effect sizes, p-values, etc.).\n",
    "\n",
    "> **Note:** <br>\n",
    "> The `\"inph\"` files are the phenotype information<br>\n",
    "> The `\"inco\"` files are the covariate information, `\"bt\"` for binary traits and `\"qt\"` for quantitative traits<br>\n",
    "> `\"bt\"` for binary traits and \"qt\" for quantitative traits<br>\n",
    "> **The analysis needs to be separated by sex**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a77eb9e-5765-4750-8708-64863aca6b1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Set parameters\n",
    "# Set genotype data\n",
    "# Adjust acording to your data\n",
    "# The \"inge\" needs to be the plink sufix files\n",
    "inge=\"path/cohort_name\"\n",
    "\n",
    "# The \"inph\" files are the phenotype information, \"bt\" for binary traits and \"qt\" for quantitative traits, separated by sex\n",
    "# Set phenotype and covariates files for females\n",
    "female_inph_bt=\"path/cohort_female_pheno_bt_forregenie.txt\"\n",
    "female_inph_qt=\"path/cohort_female_pheno_qt_forregenie.txt\"\n",
    "female_inco=\"path/cohort_female_covar_forregenie.txt\"\n",
    "\n",
    "# Set phenotype and covariates files for males\n",
    "male_inph_bt=\"path/cohort_male_pheno_bt_forregenie.txt\"\n",
    "male_inph_qt=\"path/cohort_male_pheno_qt_forregenie.txt\"\n",
    "male_inco=\"path/cohort_male_covar_forregenie.txt\"\n",
    "\n",
    "# Set output for step 1\n",
    "out_step1_qt_female=\"path/cohort/out_step1_qt_female\"\n",
    "out_step1_bt_female=\"path/cohort/out_step1_bt_female\"\n",
    "out_step1_qt_male=\"path/cohort/out_step1_qt_male\"\n",
    "out_step1_bt_male=\"path/cohort/out_step1_bt_male\"\n",
    "\n",
    "# Set output for step 2\n",
    "out_step2_qt_female=\"path/cohort/out_step2_qt_female\"\n",
    "out_step2_bt_female=\"path/cohort/out_step2_bt_female\"\n",
    "out_step2_qt_male=\"path/cohort/out_step2_qt_male\"\n",
    "out_step2_bt_male=\"path/cohort/out_step2_bt_male\"\n",
    "\n",
    "# Running for females for quantitative traits\n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${female_inph_qt} \\\n",
    "  --covarFile ${female_inco} \\\n",
    "  --qt \\\n",
    "  --pred ${out_step1_qt_female}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_qt_female}\n",
    "\n",
    "# Running for females for binary traits\n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${female_inph_bt} \\\n",
    "  --covarFile ${female_inco} \\\n",
    "  --bt \\\n",
    "  --firth --approx --pThresh 0.01 \\\n",
    "  --pred ${out_step1_bt_female}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_bt_female}\n",
    "\n",
    "# Running for males for quantitative traits\n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${male_inph_qt} \\\n",
    "  --covarFile ${male_inco} \\\n",
    "  --qt \\\n",
    "  --pred ${out_step1_qt_male}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_qt_male}\n",
    "\n",
    "# Running for males for binary traits\n",
    "regenie \\\n",
    "  --step 2 \\\n",
    "  --bgen ${inge}.bgen \\\n",
    "  --ref-first \\\n",
    "  --sample ${inge}.sample \\\n",
    "  --phenoFile ${male_inph_bt} \\\n",
    "  --covarFile ${male_inco} \\\n",
    "  --bt \\\n",
    "  --firth --approx --pThresh 0.01 \\\n",
    "  --pred ${out_step1_bt_male}_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --out ${out_step2_bt_male}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94f1fd5b-94b8-44e9-b52c-faaa56eb4ead",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
