{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eab71530-93b3-48ab-ae91-012279b4e729",
   "metadata": {},
   "outputs": [],
   "source": [
    "## First, we will perform an LD pruning with plink2\n",
    "# Substitute for the path and file name prefix of plink2 files\n",
    "in=\"/path_to_your_data/plink_prefix\"\n",
    "# Perform first pass of LD pruning\n",
    "plink2 --pfile ${in}\\\n",
    "    --indep-pairwise 50 5 0.2 --geno 0.01 --mind 0.01 --maf 0.05\\\n",
    "    --out ${in}_indep_snps\n",
    "plink2 --pfile ${in}\\\n",
    "    --extract ${in}_indep_snps.prune.in --make-bed\\\n",
    "    --out ${in}_forpcair"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55c51878-2c76-49b1-99a9-35edffb41270",
   "metadata": {},
   "source": [
    "### 2. Creating GDS files\n",
    "\n",
    "***Description:***\n",
    "\n",
    "This script will create gds files used for perform the Principal Components calculation, on the next step"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47db38e0-d34e-4a78-8888-4e4fc38de01b",
   "metadata": {},
   "source": [
    "### 2.1 Creating the script for create the gds files \n",
    "name of script: **create_gds.Rscript**\n",
    "\n",
    "You can download this script on github [LINK HERE](https://github.com/ormondr/Smoking_GWAS_LAGC/blob/main/English/02PCair/create_gds.Rscript)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd9f3135-8e24-46d5-9273-37ea7ad5d9a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################################################\n",
    "# This script uses a plink file to create a GDS file for input to PCAiR\n",
    "####################################################################################\n",
    "# set parameters\n",
    "# this function uses the library optparse to add arguments to the script\n",
    "# adding arguments to the script\n",
    "library(optparse) \n",
    "option_list = list(\n",
    "    make_option(c(\"--plinkfile\"), type=\"character\", default=NULL,\n",
    "                help=\"path to the directory storing the plink files, without plink extensions *bed/*bim/*fam\", metavar=\"character\"),\n",
    "    make_option(c(\"--out\"), type=\"character\", default=NULL,\n",
    "                help=\"output file name for GDS file, use complete path, example /data/analsyis/analysis.gds\", metavar=\"character\")\n",
    ");\n",
    "\n",
    "opt_parser = OptionParser(option_list=option_list);\n",
    "opt = parse_args(opt_parser);\n",
    "\n",
    "if (is.null(opt$plinkfile)){\n",
    "    print_help(opt_parser)\n",
    "    stop(\"At least one argument must be supplied (input file)\", call.=FALSE)\n",
    "}\n",
    "\n",
    "###################################################################################\n",
    "# load libraries\n",
    "library(GENESIS)\n",
    "library(GWASTools)\n",
    "library(SNPRelate)\n",
    "\n",
    "# create list of files\n",
    "bed_f=paste0(opt$plinkfile,\".bed\",sep=\"\")\n",
    "bim_f=paste0(opt$plinkfile,\".bim\",sep=\"\")\n",
    "fam_f=paste0(opt$plinkfile,\".fam\",sep=\"\")\n",
    "\n",
    "# create gds output files\n",
    "snpgdsBED2GDS(bed.fn=bed_f,\n",
    "              bim.fn=bim_f,\n",
    "              fam.fn=fam_f,\n",
    "              family=TRUE,\n",
    "              out.gdsfn=opt$out)\n",
    "# end of script"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d28a16f7-0640-4db5-8b8a-e8c3a79bf6f0",
   "metadata": {},
   "source": [
    "### 2.2 Create the gds files \n",
    "***Description:***\n",
    "This script will run the script \"create_gds.Rscript\", create in the previous step <br>\n",
    "Obs: needs to modifie the parameters for your files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "641df291-dcad-4c15-bfda-2e6fc0a2d87f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating GDS files\n",
    "# set parameters\n",
    "Rscript='/path_to_your_data/create_gds.Rscript'\n",
    "inpath='/path_to_your_data/01plinkldpruned'\n",
    "oupath='/path_to_your_data/02gdsprunned'\n",
    "\n",
    "# create job list file\n",
    "do Rscript ${Rscript} --plinkfile=${inpath}/$cohort_name_hapmap_metalprs_allchrs --out=${oupath}/cohort_name_hapmap_metalprs_allchrs.gds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efb5fc97-d874-4612-9470-d7e30ac7a088",
   "metadata": {},
   "source": [
    "### 3. Principal Components Calculation Using PC-AiR\n",
    "\n",
    "**Description:**\n",
    "\n",
    "This script performs Principal Component Analysis (PCA), which will be used as covariates in the GWAS.  \n",
    "The PCA is conducted using the `PC-AiR` function from the **GENESIS** package.\n",
    "\n",
    "- To learn more about PC-AiR and PC-Relate, see this vignette: [GENESIS PCA Guide](https://bioconductor.org/packages/devel/bioc/vignettes/GENESIS/inst/doc/pcair.html)  \n",
    "  - See **Section 3** for PC-AiR and **Section 4** for PC-Relate.\n",
    "  \n",
    "> **Note:**  \n",
    "> PCA requires prior LD pruning, which is performed in the first step of this pipeline.  \n",
    "> The file containing the top 10 PCs will be used for all GWAS methods (e.g., Regenie, GMMAT, Saige).  \n",
    "> The GRM file is required only for GMMAT, but we recommend generating it anyway — this ensures you won’t need to repeat this step if you decide to use GMMAT later.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53330f39-6a90-46b6-a60e-f9c0a1f7e1ab",
   "metadata": {},
   "source": [
    "### 3.1. Developing the script for PCA and GRM with PCAiR and PCRelate\n",
    "*Script written in R.*\n",
    "developing the script for PCA and GRM with PCAiR and PCRelate <br>\n",
    "this script don't need to be modified, the paths and file names will be added on the next step<br>\n",
    "name of script: **create_grm.Rscript**\n",
    "\n",
    "You can download this script on github [LINK HERE](https://github.com/ormondr/Smoking_GWAS_LAGC/blob/main/English/02PCair/create_grm.Rscript)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e61253a-6303-460c-8bbc-0026c82498ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################################################################\n",
    "# R script for creating GRM and PCs\n",
    "# day: 25 July 2025\n",
    "# author: Rafaella Ormond and Jose Jaime Martinez-Magana\n",
    "####################################################################################\n",
    "# This script uses a gds file to estimate PCs and GRM using GENESIS\n",
    "####################################################################################\n",
    "# set parameters\n",
    "# this function uses the library optparse to add arguments to the script\n",
    "# adding arguments to the script\n",
    "library(optparse) \n",
    "option_list = list(\n",
    "    make_option(c(\"--gdsfile\"), type=\"character\", default=NULL,\n",
    "                help=\"path to gds file\", metavar=\"character\"),\n",
    "    make_option(c(\"--out\"), type=\"character\", default=NULL,\n",
    "                help=\"output path for PCAiR objects, use a complete directory path and a potential name of files. Example: /home/user/out_pca\", metavar=\"character\")\n",
    ");\n",
    "\n",
    "opt_parser = OptionParser(option_list=option_list);\n",
    "opt = parse_args(opt_parser);\n",
    "\n",
    "if (is.null(opt$gdsfile)){\n",
    "  print_help(opt_parser)\n",
    "  stop(\"At least one argument must be supplied (input file)\", call.=FALSE)\n",
    "}\n",
    "\n",
    "###################################################################################\n",
    "# load libraries\n",
    "library(GENESIS)\n",
    "library(GWASTools)\n",
    "library(SNPRelate)\n",
    "library(SeqArray)\n",
    "library(parallel)\n",
    "library(BiocParallel)\n",
    "library(stringi);\n",
    "\n",
    "# stablishing a seed\n",
    "set.seed(1000)\n",
    "# stablishing cores\n",
    "cores=detectCores()\n",
    "\n",
    "# create function for LD prunning\n",
    "ld_prun=function(gds){\n",
    "snpset=snpgdsLDpruning(gds,\n",
    "                       method=\"corr\",\n",
    "                       slide.max.bp=10e6,\n",
    "                       ld.threshold=sqrt(0.1),\n",
    "                       maf=0.01,\n",
    "                       missing.rate=0.01,\n",
    "                       verbose=TRUE, num.thread=cores);\n",
    "    pruned=unlist(snpset,\n",
    "                  use.names=FALSE);\n",
    "return(pruned)\n",
    "}\n",
    "\n",
    "# create function for KING-robust analysis\n",
    "king_mat=function(gds){\n",
    "samp.id=read.gdsn(index.gdsn(gds, \"sample.id\"))\n",
    "ibd.robust=snpgdsIBDKING(gds, sample.id=samp.id, family.id=NULL, maf=0.01,missing.rate=0.01,num.thread=cores)\n",
    "return(ibd.robust)\n",
    "}\n",
    "\n",
    "\n",
    "# create function for PCAiR\n",
    "pcair_r=function(gds_geno, pruned, KINGmat){\n",
    "pcair=pcair(gds_geno, snp.include=pruned,\n",
    "            kinobj=KINGmat, divobj=KINGmat)\n",
    "return(pcair)\n",
    "}\n",
    "\n",
    "# create function for PCAiR with sample filter\n",
    "pcair_r_fil=function(gds_geno, pruned, KINGmat, sample_list){\n",
    "    pcair=pcair(gds_geno, snp.include=pruned,\n",
    "                sample.include=sample_list,\n",
    "                kinobj=KINGmat, divobj=KINGmat)\n",
    "                return(pcair)\n",
    "                }\n",
    "\n",
    "\n",
    "# create function to match the sample ID in the GDS with the sample IDs of the supplied sampleID filter\n",
    "filter_samples=function(gds_samples, incl_samples){\n",
    "    matches=unique(grep(paste(incl_samples$SampleID,collapse=\"|\"),\n",
    "                        gds_samples, value=TRUE))\n",
    "    return(matches)\n",
    "}\n",
    "\n",
    "\n",
    "# run analysis for pcs with relationships\n",
    "# open the gds object\n",
    "gds=snpgdsOpen(opt$gdsfile);\n",
    "# LD prunning\n",
    "pruned=ld_prun(gds)\n",
    "# build KING matrix\n",
    "KINGmat=king_mat(gds)\n",
    "# adjust KING matrix\n",
    "KINGmat_m=KINGmat$kinship\n",
    "# add sampleIDs to colnames y row names\n",
    "colnames(KINGmat_m)=KINGmat$sample.id\n",
    "rownames(KINGmat_m)=KINGmat$sample.id\n",
    "# get samples in gds\n",
    "gds_samples=read.gdsn(index.gdsn(gds, \"sample.id\"))\n",
    "# close the gds object\n",
    "snpgdsClose(gds)\n",
    "\n",
    "# read the gds object\n",
    "gds=GdsGenotypeReader(filename=opt$gdsfile)\n",
    "\n",
    "# create a GenotypeData class object\n",
    "gds_geno=GenotypeData(gds)\n",
    "\n",
    "# run PCAiR\n",
    "PCair=pcair_r(gds_geno, pruned, KINGmat_m)\n",
    "PCairpart=pcairPartition(kinobj = KINGmat_m, divobj = KINGmat_m)\n",
    "\n",
    "# PC relate calculation \n",
    "gdsData=GenotypeBlockIterator(gds_geno, snpInclude=pruned)\n",
    "PCrelate=pcrelate(gdsData,\n",
    "                  pcs=PCair$vectors[,1:2],\n",
    "                  training.set=PCairpart$unrels,\n",
    "                  BPPARAM=BiocParallel::SerialParam())\n",
    "\n",
    "# making GRM\n",
    "grm=pcrelateToMatrix(PCrelate)\n",
    "# making sparce matrix\n",
    "grm[grm<0.05]=0\n",
    "grm_sparse=as.matrix(grm, sparse = TRUE)\n",
    "\n",
    "# save file\n",
    "out=list()\n",
    "out$PCair=PCair\n",
    "out$PCairpart=PCairpart\n",
    "out$KINGmat=KINGmat_m\n",
    "out$pruned=pruned\n",
    "out$PCrelate=PCrelate\n",
    "out$grm=grm\n",
    "out$grm_sparse=grm_sparse\n",
    "saveRDS(file=paste0(opt$out,\".rds\",\"\"),out)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "072ce1d8-71f4-4003-8daa-ffbaa6745556",
   "metadata": {},
   "source": [
    "### 3.2. Generate the PCA and GRM files.\n",
    "***Description:***\n",
    "\n",
    "This sctipt will run the \"create_grm.Rscript\" created in the previous step.\n",
    "You will need to adjust this step according to your own input files.\n",
    "The output will be a .RDS file containing both the PCA information and the Genetic Relationship Matrix (GRM)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e68639b-7e4b-41db-b674-25823fdcb838",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run R script for creating GDS files\n",
    "# add the path of the create_grm.Rscript from the previous step\n",
    "Rscript=\"create_grm.Rscript\"\n",
    "Rscript $Rscript \\\n",
    "--gdsfile=gdsfile_prunned \\\n",
    "--out=cohort_name.gds_prunned_grm_pca"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d9f7e11-1d7d-4881-bc6a-9a30a1d2368b",
   "metadata": {},
   "source": [
    "### 3.3. Save the Pca file\n",
    "\n",
    "***Description:***\n",
    "This script extracts the top 10 principal components from the .RDS file generated in the previous step and saves them in a separate file to be used as covariates in Regenie."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "001a5026-f1c5-4b23-a249-b367101a4f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the PCA/GRM result file\n",
    "pca_grm <- readRDS(\"cohort_name.gds_prunned_grm_pca.rds\")\n",
    "\n",
    "# Extract sample IDs (IIDs)\n",
    "iids <- rownames(pca_grm$PCair$vectors)\n",
    "\n",
    "# Extract top 10 PCs\n",
    "pcs <- pca_grm$PCair$vectors[,c(1:10)]\n",
    "pcs <- as.data.frame(pcs)\n",
    "colnames(pcs) <- paste0(rep(\"PC\",10),rep(1:10))\n",
    "\n",
    "# add sample ID\n",
    "pcs$SampleID=rownames(pcs)\n",
    "\n",
    "covariate_table <- pcs\n",
    "\n",
    "# Save to file\n",
    "write.table(covariate_table,\n",
    "            file = \"cohort_10pcs_forregenie.txt\",\n",
    "            quote = FALSE,\n",
    "            row.names = FALSE,\n",
    "            col.names = TRUE,\n",
    "            sep = \"\\t\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
