{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6202ce00-100b-4ea0-8e0a-ef88bca29f6a",
   "metadata": {},
   "source": [
    "# Principal Components Inference\n",
    "\n",
    "This script performs Principal Component Analysis (PCA) to generate covariates for the GWAS of smoking traits in the LAGC cohorts.  \n",
    "The following steps will be carried out:\n",
    "\n",
    "1) LD pruning using PLINK  \n",
    "2) Principal Components calculation using PC-AiR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6404901-caff-48ff-85af-3f0eea9be1ce",
   "metadata": {},
   "source": [
    "### 1) LD pruning using PLINK\n",
    "\n",
    "Description: This code is for perform the LD pruning usinh PLINK2 for PCs analysis\n",
    "\n",
    "You also can do the transformation for bfiles, but if the files are slited by chromossomes, with pfiles is faster to merge those\n",
    "\n",
    "We can download Plink version 1.9 and version 2.0 following the steps from their website.<br>\n",
    "For instal plink2 [access here](https://www.cog-genomics.org/plink/2.0/)<br>\n",
    "For instal plink1.9 [access here](https://www.cog-genomics.org/plink/1.9/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eab71530-93b3-48ab-ae91-012279b4e729",
   "metadata": {},
   "outputs": [],
   "source": [
    "## First, we will perform an LD pruning with plink2\n",
    "# Substitute for the path and file name prefix of plink2 files\n",
    "in=\"path_and_file\"\n",
    "# Perform first pass of LD pruning\n",
    "# MxGDAR-Fz1\n",
    "plink2 --pfile ${in}\\\n",
    "    --indep-pairwise 50 5 0.2 --geno 0.01 --mind 0.01 --maf 0.05\\\n",
    "    --out ${in}_indep_snps\n",
    "plink2 --pfile ${in}\\\n",
    "    --extract ${in}_indep_snps.prune.in --make-bed\\\n",
    "    --out ${in}_forpcair"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efb5fc97-d874-4612-9470-d7e30ac7a088",
   "metadata": {},
   "source": [
    "### 2) Principal Components Calculation Using PC-AiR\n",
    "\n",
    "**Description:**\n",
    "\n",
    "This script performs Principal Components (PCs) analysis, which will be used as covariates in the GWAS.  \n",
    "The PCA is performed using the `PC-AiR` function from the **GENESIS** package.\n",
    "\n",
    "To download **GENESIS**, follow this link: [Access here](https://www.bioconductor.org/packages/release/bioc/html/GENESIS.html)\n",
    "\n",
    "> **Note:** Principal Component Analysis requires prior LD pruning, which is performed in the first step of this file.\n",
    "\n",
    "*Script written in R.*\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c643479c-1779-4be8-a966-25649b681d03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load libraries\n",
    "library(GENESIS)\n",
    "library(GWASTools)\n",
    "library(SNPRelate)\n",
    "library(SeqArray)\n",
    "library(parallel)\n",
    "library(BiocParallel);\n",
    "\n",
    "############################################################################################\n",
    "## !!! This is the only section that needs adjustment based on your data !!! ##\n",
    "## Set parameters\n",
    "# Change directory to the directory where the PLINK LD pruned files are stored\n",
    "wkdir=\"path_plinkfiles/\"\n",
    "setwd(wkdir)\n",
    "# Create a list of files for input to PCAiR\n",
    "filename=\"cohortname_forpcair\"\n",
    "bed_f=paste0(filename,\".bed\",sep=\"\")\n",
    "bim_f=paste0(filename,\".bim\",sep=\"\")\n",
    "fam_f=paste0(filename,\".fam\",sep=\"\")\n",
    "\n",
    "# Create output files for a GDS file, keep the GDS extension\n",
    "gds_output=\"cohortname_merged_forpcair.gds\"\n",
    "\n",
    "# Create a file name to store the PCA output, keep the CSV extension\n",
    "outfile_name=\"cohortname_pcs32.csv\"\n",
    "\n",
    "# create gds output files\n",
    "snpgdsBED2GDS(bed.fn=bed_f,\n",
    "              bim.fn=bim_f,\n",
    "              fam.fn=fam_f,\n",
    "              family=TRUE,\n",
    "              out.gdsfn=gds_output)\n",
    "####################################################################################\n",
    "# establishing a seed\n",
    "set.seed(1000)\n",
    "# establishing cores\n",
    "cores=detectCores()\n",
    "####################################################################################\n",
    "# Create functions\n",
    "# Create a function for LD pruning\n",
    "ld_prun = function(gds){\n",
    "snpset=snpgdsLDpruning(gds,\n",
    "                       method=\"corr\",\n",
    "                       slide.max.bp=10e6,\n",
    "                       ld.threshold=sqrt(0.1),\n",
    "                       maf=0.01,\n",
    "                       missing.rate=0.01,\n",
    "                       verbose=TRUE, num.thread=cores);\n",
    "    pruned=unlist(snpset,\n",
    "                  use.names=FALSE);\n",
    "return(pruned)\n",
    "}\n",
    "\n",
    "# Create a function for KING-robust analysis\n",
    "king_mat = function(gds){\n",
    "samp.id=read.gdsn(index.gdsn(gds, \"sample.id\"))\n",
    "ibd.robust=snpgdsIBDKING(gds, sample.id=samp.id, family.id=NULL, maf=0.01,missing.rate=0.01,num.thread=cores)\n",
    "return(ibd.robust)\n",
    "}\n",
    "\n",
    "# Create a function for PCAiR\n",
    "pcair_r = function(gds_geno, pruned, KINGmat){\n",
    "pcair=pcair(gds_geno, snp.include=pruned,\n",
    "            kinobj=KINGmat, divobj=KINGmat)\n",
    "return(pcair)\n",
    "}\n",
    "####################################################################################\n",
    "# set input data\n",
    "gds=gds_output\n",
    "\n",
    "# run analysis for PCs with relationships\n",
    "# open the GDS object\n",
    "gds=snpgdsOpen(gds);\n",
    "# LD pruning\n",
    "pruned=ld_prun(gds)\n",
    "# build KING matrix\n",
    "KINGmat=king_mat(gds)\n",
    "# adjust KING matrix\n",
    "KINGmat_m=KINGmat$kinship\n",
    "# add sampleIDs to colnames y row names\n",
    "colnames(KINGmat_m)=KINGmat$sample.id\n",
    "rownames(KINGmat_m)=KINGmat$sample.id\n",
    "# get samples in gds\n",
    "gds_samples=read.gdsn(index.gdsn(gds, \"sample.id\"))\n",
    "# close the gds object\n",
    "snpgdsClose(gds)\n",
    "####################################################################################\n",
    "# Warning: review this step!!!!!\n",
    "# read the GDS object\n",
    "gds=GdsGenotypeReader(filename=gds_output)\n",
    "\n",
    "# create a GenotypeData class object\n",
    "gds_geno=GenotypeData(gds)\n",
    "\n",
    "# run PCAiR\n",
    "PCair=pcair_r(gds_geno, pruned, KINGmat_m)\n",
    "\n",
    "# Create a file of PCs\n",
    "PCair_df=as.data.frame(PCair$vectors)\n",
    "# assign colnames\n",
    "colnames(PCair_df)=paste0(\"PC\",rep(1:32))\n",
    "# add IID\n",
    "PCair_df$IID=row.names(PCair_df)\n",
    "\n",
    "# save file\n",
    "write.csv(file=outfile_name,\n",
    "         PCair_df,\n",
    "         quote=FALSE,\n",
    "         row.names=FALSE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
